--- a/net/minecraft/client/renderer/ItemRenderer.java
+++ b/net/minecraft/client/renderer/ItemRenderer.java
@@ -183,9 +183,10 @@
 
                 if (stack.hasEffect() && (!Config.isCustomItems() || !CustomItems.renderCustomEffect(this, stack, model)))
                 {
+                	final IBakedModel modelF = model;
                     renderEffect(this.textureManager, () ->
                     {
-                        this.renderModel(model, -8372020);
+                        this.renderModel(modelF, -8372020);
                     }, 8);
                 }
             }
