--- a/net/minecraft/client/renderer/WorldRenderer.java
+++ b/net/minecraft/client/renderer/WorldRenderer.java
@@ -859,7 +859,7 @@
             boolean flag2 = Reflector.IForgeTileEntity_getRenderBoundingBox.exists();
             label192:
 
-            for (WorldRenderer.LocalRenderInformationContainer worldrenderer$localrenderinformationcontainer : this.renderInfosTileEntities)
+            for (WorldRenderer.LocalRenderInformationContainer worldrenderer$localrenderinformationcontainer : (List<WorldRenderer.LocalRenderInformationContainer>)this.renderInfosTileEntities)
             {
                 List<TileEntity> list2 = worldrenderer$localrenderinformationcontainer.renderChunk.getCompiledChunk().getTileEntities();
 
