--- a/net/minecraft/client/gui/inventory/GuiContainer.java
+++ b/net/minecraft/client/gui/inventory/GuiContainer.java
@@ -133,9 +133,7 @@
                 GlStateManager.disableDepthTest();
                 int j1 = slot.xPos;
                 int k1 = slot.yPos;
-                GlStateManager.colorMask(true, true, true, false);
                 this.drawGradientRect(j1, k1, j1 + 16, k1 + 16, -2130706433, -2130706433);
-                GlStateManager.colorMask(true, true, true, true);
                 GlStateManager.enableLighting();
                 GlStateManager.enableDepthTest();
             }
@@ -420,7 +418,7 @@
                             }
                             else
                             {
-                                boolean flag2 = l != -999 && (InputMappings.isKeyDown(340) || InputMappings.isKeyDown(344));
+                                boolean flag2 = l != -999 && (InputMappings.isKeyDown(340) || InputMappings.isKeyDown(344) || isFakeShift());
                                 ClickType clicktype = ClickType.PICKUP;
 
                                 if (flag2)
