diff -r -U 3 minecraft\net\minecraft\tags\NetworkTagCollection.java minecraft_patched\net\minecraft\tags\NetworkTagCollection.java
--- minecraft\net\minecraft\tags\NetworkTagCollection.java
+++ minecraft_patched\net\minecraft\tags\NetworkTagCollection.java
@@ -1,6 +1,8 @@
 package net.minecraft.tags;
 
 import com.google.common.collect.Lists;
+
+import java.util.Collection;
 import java.util.List;
 import java.util.Map.Entry;
 import java.util.function.Function;
@@ -43,7 +45,7 @@
             list.add(this.field_200044_a.func_148754_a(p_200043_1_.func_150792_a()));
          }
 
-         this.func_200039_c().put(resourcelocation, Tag.Builder.func_200047_a().func_200046_a(list).func_200051_a(resourcelocation));
+         this.func_200039_c().put(resourcelocation, (Tag<T>) Tag.Builder.func_200047_a().func_200046_a((Collection<Object>) list).func_200051_a(resourcelocation));
       }
 
    }
