diff -r -U 3 minecraft\net\minecraft\world\WorldServer.java minecraft_patched\net\minecraft\world\WorldServer.java
--- minecraft\net\minecraft\world\WorldServer.java
+++ minecraft_patched\net\minecraft\world\WorldServer.java
@@ -57,6 +57,7 @@
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.IProgressUpdate;
 import net.minecraft.util.IThreadListener;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.WeightedRandom;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
@@ -108,7 +108,7 @@
 
    public WorldServer(MinecraftServer p_i26_1_, ISaveHandler p_i26_2_, WorldSavedDataStorage p_i26_3_, WorldInfo p_i26_4_, DimensionType p_i26_5_, Profiler p_i26_6_) {
       super(p_i26_2_, p_i26_3_, p_i26_4_, p_i26_5_.func_186070_d(), p_i26_6_, false);
-      Predicate predicate = (p_205341_0_) -> {
+      Predicate<Block> predicate = (p_205341_0_) -> {
          return p_205341_0_ == null || p_205341_0_.func_176223_P().func_196958_f();
       };
       IRegistry iregistry1 = IRegistry.field_501011_g;
@@ -116,8 +116,8 @@
       Function function = iregistry1::func_501993_b;
       IRegistry iregistry2 = IRegistry.field_501011_g;
       IRegistry.field_501011_g.getClass();
-      this.field_94579_S = new ServerTickList<Block>(this, predicate, function, iregistry2::func_501994_a, this::func_205338_b);
-      predicate = (p_205774_0_) -> {
+      this.field_94579_S = new ServerTickList<Block>(this, predicate, iregistry1::func_501993_b, IRegistry.field_501011_g::func_501994_a, this::func_205338_b);
+      Predicate<Fluid> predicateFluid = (p_205774_0_) -> {
          return p_205774_0_ == null || p_205774_0_ == Fluids.field_204541_a;
       };
       IRegistry iregistry = IRegistry.field_501012_h;
@@ -125,7 +125,7 @@
       Function function1 = iregistry::func_501993_b;
       iregistry2 = IRegistry.field_501012_h;
       IRegistry.field_501012_h.getClass();
-      this.field_205342_P = new ServerTickList<Fluid>(this, predicate, function1, iregistry2::func_501994_a, this::func_205339_a);
+      this.field_205342_P = new ServerTickList<Fluid>(this, predicateFluid, iregistry::func_501993_b, IRegistry.field_501012_h::func_501994_a, this::func_205339_a);
       this.field_175740_d = new VillageSiege(this);
       this.field_147490_S = new ObjectLinkedOpenHashSet<BlockEventData>();
       this.field_73061_a = p_i26_1_;
@@ -264,7 +264,7 @@
    protected void func_73053_d() {
       this.field_73068_P = false;
 
-      for(EntityPlayer entityplayer : (List)this.field_73010_i.stream().filter(EntityPlayer::func_70608_bn).collect(Collectors.toList())) {
+      for(EntityPlayer entityplayer : this.field_73010_i.stream().filter(EntityPlayer::func_70608_bn).collect(Collectors.toList())) {
          entityplayer.func_70999_a(false, false, true);
       }
 
