diff -r -U 3 minecraft\net\minecraft\client\renderer\texture\TextureManager.java minecraft_patched\net\minecraft\client\renderer\texture\TextureManager.java
--- minecraft\net\minecraft\client\renderer\texture\TextureManager.java
+++ minecraft_patched\net\minecraft\client\renderer\texture\TextureManager.java
@@ -65,8 +65,9 @@
          CrashReport crashreport = CrashReport.func_85055_a(throwable, "Registering texture");
          CrashReportCategory crashreportcategory = crashreport.func_85058_a("Resource location being registered");
          crashreportcategory.func_71507_a("Resource location", p_110579_1_);
+         final ITextureObject p_110579_2_F = p_110579_2_;
          crashreportcategory.func_189529_a("Texture object class", () -> {
-            return p_110579_2_.getClass().getName();
+            return p_110579_2_F.getClass().getName();
          });
          throw new ReportedException(crashreport);
       }
