diff -r -U 3 minecraft\net\minecraft\world\chunk\ChunkPrimer.java minecraft_patched\net\minecraft\world\chunk\ChunkPrimer.java
--- minecraft\net\minecraft\world\chunk\ChunkPrimer.java
+++ minecraft_patched\net\minecraft\world\chunk\ChunkPrimer.java
@@ -25,6 +25,7 @@
 import net.minecraft.init.Fluids;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.registry.RegistryNamespacedDefaultedByKey;
@@ -79,7 +79,7 @@
       this.field_205769_s = Maps.<GenerationStage.Carving, BitSet>newHashMap();
       this.field_201654_b = p_i48700_1_;
       this.field_201668_p = p_i48700_2_;
-      Predicate predicate = (p_205332_0_) -> {
+      Predicate<Block> predicate = (p_205332_0_) -> {
          return p_205332_0_ == null || p_205332_0_.func_176223_P().func_196958_f();
       };
       IRegistry iregistry1 = IRegistry.field_501011_g;
@@ -87,8 +87,8 @@
       Function function = iregistry1::func_501993_b;
       IRegistry iregistry2 = IRegistry.field_501011_g;
       IRegistry.field_501011_g.getClass();
-      this.field_201664_l = new ChunkPrimerTickList<Block>(predicate, function, iregistry2::func_501994_a, p_i48700_1_);
-      predicate = (p_205766_0_) -> {
+      this.field_201664_l = new ChunkPrimerTickList<Block>(predicate, iregistry1::func_501993_b, IRegistry.field_501011_g::func_501994_a, p_i48700_1_);
+      Predicate<Fluid> predicateFluid = (p_205766_0_) -> {
          return p_205766_0_ == null || p_205766_0_ == Fluids.field_204541_a;
       };
       IRegistry iregistry = IRegistry.field_501012_h;
@@ -96,7 +96,7 @@
       Function function1 = iregistry::func_501993_b;
       iregistry2 = IRegistry.field_501012_h;
       IRegistry.field_501012_h.getClass();
-      this.field_205333_q = new ChunkPrimerTickList<Fluid>(predicate, function1, iregistry2::func_501994_a, p_i48700_1_);
+      this.field_205333_q = new ChunkPrimerTickList<Fluid>(predicateFluid, iregistry::func_501993_b, IRegistry.field_501012_h::func_501994_a, p_i48700_1_);
    }
 
    public static ShortList func_205330_a(ShortList[] p_205330_0_, int p_205330_1_) {
