diff -r -U 3 minecraft\net\minecraft\util\datafix\fixes\KeyOptionsTranslation.java minecraft_patched\net\minecraft\util\datafix\fixes\KeyOptionsTranslation.java
--- minecraft\net\minecraft\util\datafix\fixes\KeyOptionsTranslation.java
+++ minecraft_patched\net\minecraft\util\datafix\fixes\KeyOptionsTranslation.java
@@ -30,7 +30,7 @@
 
                   return Pair.of(p_209666_1_.getKey(), p_209666_1_.getValue());
                }).collect(Collectors.toMap(Pair::getFirst, Pair::getSecond)));
-            }).orElse(p_209668_0_);
+            }).orElse((Dynamic)p_209668_0_);
          });
       });
    }
