diff -r -U 3 minecraft\net\minecraft\util\datafix\fixes\BookPagesStrictJSON.java minecraft_patched\net\minecraft\util\datafix\fixes\BookPagesStrictJSON.java
--- minecraft\net\minecraft\util\datafix\fixes\BookPagesStrictJSON.java
+++ minecraft_patched\net\minecraft\util\datafix\fixes\BookPagesStrictJSON.java
@@ -72,7 +72,7 @@
             });
          });
          p_209633_1_.getClass();
-         return (Dynamic)DataFixUtils.orElse(optional.map(p_209633_1_::createList), p_209633_1_.emptyList());
+         return (Dynamic)DataFixUtils.orElse(optional.map((x)->p_209633_1_.createList((java.util.stream.Stream)x)), p_209633_1_.emptyList());
       });
    }
 