diff -r -U 3 minecraft\net\minecraft\util\datafix\fixes\ObjectiveDisplayName.java minecraft_patched\net\minecraft\util\datafix\fixes\ObjectiveDisplayName.java
--- minecraft\net\minecraft\util\datafix\fixes\ObjectiveDisplayName.java
+++ minecraft_patched\net\minecraft\util\datafix\fixes\ObjectiveDisplayName.java
@@ -33,7 +33,7 @@
                         return ITextComponent.Serializer.func_150696_a(new TextComponentString(p_211865_0_));
                      });
                      p_211861_0_.getClass();
-                     return (Dynamic)DataFixUtils.orElse(optional.map(p_211861_0_::createString), p_211864_1_);
+                     return (Dynamic)DataFixUtils.orElse(optional.map((x)->p_211861_0_.createString((String)x)), p_211864_1_);
                   });
                });
             };
