diff -r -U 3 minecraft\net\minecraft\util\datafix\fixes\TrappedChest.java minecraft_patched\net\minecraft\util\datafix\fixes\TrappedChest.java
--- minecraft\net\minecraft\util\datafix\fixes\TrappedChest.java
+++ minecraft_patched\net\minecraft\util\datafix\fixes\TrappedChest.java
@@ -36,7 +36,7 @@
          throw new IllegalStateException("Tile entity type is not a list type.");
       } else {
          ListType<?> listtype = (ListType)type2;
-         OpticFinder<? extends List<?>> opticfinder = DSL.<List<?>>fieldFinder("TileEntities", listtype);
+         OpticFinder<? extends List<?>> opticfinder = DSL.fieldFinder("TileEntities", listtype);
          Type<?> type3 = this.getInputSchema().getType(TypeReferences.field_211287_c);
          OpticFinder<?> opticfinder1 = type3.findField("Level");
          OpticFinder<?> opticfinder2 = opticfinder1.type().findField("Sections");
@@ -44,7 +44,7 @@
          if (!(type4 instanceof ListType)) {
             throw new IllegalStateException("Expecting sections to be a list.");
          } else {
-            Type<?> type5 = (type4).getElement();
+            Type<?> type5 = ((ListType)type4).getElement();
             OpticFinder<?> opticfinder3 = DSL.typeFinder(type5);
             return TypeRewriteRule.seq((new AddNewChoices(this.getOutputSchema(), "AddTrappedChestFix", TypeReferences.field_211294_j)).makeRule(), this.fixTypeEverywhereTyped("Trapped Chest fix", type3, (p_501097_5_) -> {
                return p_501097_5_.updateTyped(opticfinder1, (p_501098_4_) -> {
@@ -70,7 +70,7 @@
                      Dynamic<?> dynamic = (Dynamic)p_502073_4_.get(DSL.remainderFinder());
                      int k = dynamic.getInt("xPos");
                      int l = dynamic.getInt("zPos");
-                     TaggedChoiceType<String> taggedchoicetype = this.getInputSchema().findChoiceType(TypeReferences.field_211294_j);
+                     TaggedChoiceType<String> taggedchoicetype = (TaggedChoiceType<String>)this.getInputSchema().findChoiceType(TypeReferences.field_211294_j);
                      return p_502073_4_.updateTyped(opticfinder, (p_502074_4_) -> {
                         return p_502074_4_.updateTyped(taggedchoicetype.finder(), (p_502075_4_) -> {
                            Dynamic<?> dynamic1 = (Dynamic)p_502075_4_.getOrCreate(DSL.remainderFinder());
