diff -r -U 3 minecraft\net\minecraft\client\gui\GuiLabel.java minecraft_patched\net\minecraft\client\gui\GuiLabel.java
--- minecraft\net\minecraft\client\gui\GuiLabel.java
+++ minecraft_patched\net\minecraft\client\gui\GuiLabel.java
@@ -8,15 +8,15 @@
    protected int field_146161_f;
    public int field_146162_g;
    public int field_146174_h;
-   private final List<String> field_146173_k;
+   private List<String> field_146173_k;
    private boolean field_146170_l;
    public boolean field_146172_j;
    private boolean field_146171_m;
-   private final int field_146168_n;
+   private int field_146168_n;
    private int field_146169_o;
    private int field_146166_p;
    private int field_146165_q;
-   private final FontRenderer field_146164_r;
+   private FontRenderer field_146164_r;
    private int field_146163_s;
 
    public void func_194997_a(int p_194997_1_, int p_194997_2_, float p_194997_3_) {
