diff -r -U 3 minecraft\net\minecraft\world\gen\placement\TopSolidRangeConfig.java minecraft_patched\net\minecraft\world\gen\placement\TopSolidRangeConfig.java
--- minecraft\net\minecraft\world\gen\placement\TopSolidRangeConfig.java
+++ minecraft_patched\net\minecraft\world\gen\placement\TopSolidRangeConfig.java
@@ -1,6 +1,6 @@
 package net.minecraft.world.gen.placement;
 
 public class TopSolidRangeConfig implements IPlacementConfig {
-   public final int field_204632_a;
-   public final int field_204633_b;
+   public int field_204632_a;
+   public int field_204633_b;
 }
