diff -r -U 3 minecraft\net\minecraft\stats\IStatFormater.java minecraft_patched\net\minecraft\stats\IStatFormater.java
--- minecraft\net\minecraft\stats\IStatFormater.java
+++ minecraft_patched\net\minecraft\stats\IStatFormater.java
@@ -11,6 +11,7 @@
    DecimalFormat DECIMAL_FORMAT = (DecimalFormat)Util.func_200696_a(new DecimalFormat("########0.00"), (p_lambda$static$0_0_) -> {
       p_lambda$static$0_0_.setDecimalFormatSymbols(DecimalFormatSymbols.getInstance(Locale.ROOT));
    });
+   static NumberFormat numberformat = NumberFormat.getIntegerInstance(Locale.US);
    IStatFormater DEFAULT = numberformat::format;
    IStatFormater DIVIDE_BY_TEN = (p_lambda$static$0_0_) -> {
       return DECIMAL_FORMAT.format((double)p_lambda$static$0_0_ * 0.1D);
@@ -44,7 +44,4 @@
 
    String format(int var1);
 
-   static {
-      NumberFormat numberformat = NumberFormat.getIntegerInstance(Locale.US);
-   }
 }
