diff -r -U 3 minecraft\net\minecraft\util\WorldChunkEnumerator.java minecraft_patched\net\minecraft\util\WorldChunkEnumerator.java
--- minecraft\net\minecraft\util\WorldChunkEnumerator.java
+++ minecraft_patched\net\minecraft\util\WorldChunkEnumerator.java
@@ -80,7 +80,7 @@
             return list;
          }
 
-         arraylist = list;
+         arraylist = (ArrayList) list;
       } catch (Throwable var18) {
          arraylist = Lists.newArrayList();
          return arraylist;
