diff -r -U 3 minecraft\net\minecraft\util\datafix\fixes\ChunkPaletteFormat.java minecraft_patched\net\minecraft\util\datafix\fixes\ChunkPaletteFormat.java
--- minecraft\net\minecraft\util\datafix\fixes\ChunkPaletteFormat.java
+++ minecraft_patched\net\minecraft\util\datafix\fixes\ChunkPaletteFormat.java
@@ -134,7 +134,7 @@
       super(p_i49676_1_, p_i49676_2_);
    }
 
-   private static void func_209300_a(Map<String, Dynamic<?>> p_209300_0_, int p_209300_1_, String p_209300_2_, String p_209300_3_) {
+   private static void func_209300_a(Map p_209300_0_, int p_209300_1_, String p_209300_2_, String p_209300_3_) {
       p_209300_0_.put(p_209300_1_ + "north", BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209300_2_ + "_wall_" + p_209300_3_ + "',Properties:{facing:'north'}}"));
       p_209300_0_.put(p_209300_1_ + "east", BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209300_2_ + "_wall_" + p_209300_3_ + "',Properties:{facing:'east'}}"));
       p_209300_0_.put(p_209300_1_ + "south", BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209300_2_ + "_wall_" + p_209300_3_ + "',Properties:{facing:'south'}}"));
@@ -146,7 +146,7 @@
 
    }
 
-   private static void func_209301_a(Map<String, Dynamic<?>> p_209301_0_, String p_209301_1_, int p_209301_2_) {
+   private static void func_209301_a(Map p_209301_0_, String p_209301_1_, int p_209301_2_) {
       p_209301_0_.put("minecraft:" + p_209301_1_ + "eastlowerleftfalsefalse", BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209301_1_ + "',Properties:{facing:'east',half:'lower',hinge:'left',open:'false',powered:'false'}}"));
       p_209301_0_.put("minecraft:" + p_209301_1_ + "eastlowerleftfalsetrue", BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209301_1_ + "',Properties:{facing:'east',half:'lower',hinge:'left',open:'false',powered:'true'}}"));
       p_209301_0_.put("minecraft:" + p_209301_1_ + "eastlowerlefttruefalse", BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209301_1_ + "',Properties:{facing:'east',half:'lower',hinge:'left',open:'true',powered:'false'}}"));
@@ -213,7 +213,7 @@
       p_209301_0_.put("minecraft:" + p_209301_1_ + "westupperrighttruetrue", BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209301_1_ + "',Properties:{facing:'west',half:'upper',hinge:'right',open:'true',powered:'true'}}"));
    }
 
-   private static void func_209307_a(Map<String, Dynamic<?>> p_209307_0_, int p_209307_1_, String p_209307_2_) {
+   private static void func_209307_a(Map p_209307_0_, int p_209307_1_, String p_209307_2_) {
       p_209307_0_.put("southfalsefoot" + p_209307_1_, BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209307_2_ + "_bed',Properties:{facing:'south',occupied:'false',part:'foot'}}"));
       p_209307_0_.put("westfalsefoot" + p_209307_1_, BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209307_2_ + "_bed',Properties:{facing:'west',occupied:'false',part:'foot'}}"));
       p_209307_0_.put("northfalsefoot" + p_209307_1_, BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209307_2_ + "_bed',Properties:{facing:'north',occupied:'false',part:'foot'}}"));
@@ -228,7 +228,7 @@
       p_209307_0_.put("easttruehead" + p_209307_1_, BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209307_2_ + "_bed',Properties:{facing:'east',occupied:'true',part:'head'}}"));
    }
 
-   private static void func_209297_b(Map<String, Dynamic<?>> p_209297_0_, int p_209297_1_, String p_209297_2_) {
+   private static void func_209297_b(Map p_209297_0_, int p_209297_1_, String p_209297_2_) {
       for(int i = 0; i < 16; ++i) {
          p_209297_0_.put("" + i + "_" + p_209297_1_, BlockStateFlatteningMap.func_210048_b("{Name:'minecraft:" + p_209297_2_ + "_banner',Properties:{rotation:'" + i + "'}}"));
       }
@@ -260,7 +260,7 @@
 
    private Dynamic<?> func_209712_b(Dynamic<?> p_209712_1_) {
       Optional<? extends Dynamic<?>> optional = p_209712_1_.get("Level");
-      return optional.isPresent() && ((Dynamic)optional.get()).get("Sections").flatMap(Dynamic::getStream).isPresent() ? p_209712_1_.set("Level", (new ChunkPaletteFormat.UpgradeChunk(optional.get())).func_210058_a()) : p_209712_1_;
+      return optional.isPresent() && (optional.get()).get("Sections").flatMap(Dynamic::getStream).isPresent() ? p_209712_1_.set("Level", (new ChunkPaletteFormat.UpgradeChunk(optional.get())).func_210058_a()) : p_209712_1_;
    }
 
    public TypeRewriteRule makeRule() {
