diff -r -U 3 minecraft\net\minecraft\client\renderer\texture\TextureUtil.java minecraft_patched\net\minecraft\client\renderer\texture\TextureUtil.java
--- minecraft\net\minecraft\client\renderer\texture\TextureUtil.java
+++ minecraft_patched\net\minecraft\client\renderer\texture\TextureUtil.java
@@ -117,8 +117,8 @@
          bytebuffer = MemoryUtil.memAlloc((int)filechannel.size() + 1);
 
          while(true) {
-            if (filechannel.read(bytebuffer) != -1) {
-               continue;
+            if (filechannel.read(bytebuffer) == -1) {
+               break;
             }
          }
       } else {
