diff -r -U 3 minecraft\net\minecraft\entity\monster\EntityVex.java minecraft_patched\net\minecraft\entity\monster\EntityVex.java
--- minecraft\net\minecraft\entity\monster\EntityVex.java
+++ minecraft_patched\net\minecraft\entity\monster\EntityVex.java
@@ -130,12 +130,12 @@
    }
 
    private boolean func_190656_b(int p_190656_1_) {
-      int i = this.field_70180_af.func_187225_a(field_190664_a);
+      int i = (int) this.field_70180_af.func_187225_a(field_190664_a);
       return (i & p_190656_1_) != 0;
    }
 
    private void func_190660_a(int p_190660_1_, boolean p_190660_2_) {
-      int i = this.field_70180_af.func_187225_a(field_190664_a);
+      int i = (int) this.field_70180_af.func_187225_a(field_190664_a);
       if (p_190660_2_) {
          i = i | p_190660_1_;
       } else {
