diff -r -U 3 minecraft\net\minecraft\world\chunk\ChunkPrimerWrapper.java minecraft_patched\net\minecraft\world\chunk\ChunkPrimerWrapper.java
--- minecraft\net\minecraft\world\chunk\ChunkPrimerWrapper.java
+++ minecraft_patched\net\minecraft\world\chunk\ChunkPrimerWrapper.java
@@ -15,6 +15,7 @@
 import net.minecraft.init.Fluids;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.registry.RegistryNamespacedDefaultedByKey;
@@ -177,7 +177,7 @@
    }
 
    public ChunkPrimerTickList<Block> func_205218_i_() {
-      Predicate predicate = (p_209219_0_) -> {
+      Predicate<Block> predicate = (p_209219_0_) -> {
          return p_209219_0_.func_176223_P().func_196958_f();
       };
       IRegistry iregistry = IRegistry.field_501011_g;
@@ -185,7 +185,7 @@
       Function function = iregistry::func_501993_b;
       IRegistry iregistry1 = IRegistry.field_501011_g;
       IRegistry.field_501011_g.getClass();
-      return new ChunkPrimerTickList<Block>(predicate, function, iregistry1::func_501994_a, this.func_76632_l());
+      return new ChunkPrimerTickList<Block>(predicate, iregistry::func_501993_b, IRegistry.field_501011_g::func_501994_a, this.func_76632_l());
    }
 
    public ChunkPrimerTickList<Fluid> func_205217_j_() {
@@ -197,7 +197,7 @@
       Function function = iregistry::func_501993_b;
       IRegistry iregistry1 = IRegistry.field_501012_h;
       IRegistry.field_501012_h.getClass();
-      return new ChunkPrimerTickList<Fluid>(predicate, function, iregistry1::func_501994_a, this.func_76632_l());
+      return new ChunkPrimerTickList<Fluid>(predicate, iregistry::func_501993_b, IRegistry.field_501012_h::func_501994_a, this.func_76632_l());
    }
 
    public BitSet func_205749_a(GenerationStage.Carving p_205749_1_) {
