diff -r -U 3 minecraft\net\minecraft\entity\monster\EntitySlime.java minecraft_patched\net\minecraft\entity\monster\EntitySlime.java
--- minecraft\net\minecraft\entity\monster\EntitySlime.java
+++ minecraft_patched\net\minecraft\entity\monster\EntitySlime.java
@@ -167,7 +167,7 @@
    }
 
    public EntityType<? extends EntitySlime> func_200600_R() {
-      return super.func_200600_R();
+      return (EntityType) super.func_200600_R();
    }
 
    public void func_70106_y() {
