diff -r -U 3 minecraft\net\minecraft\util\datafix\fixes\RidingToPassengers.java minecraft_patched\net\minecraft\util\datafix\fixes\RidingToPassengers.java
--- minecraft\net\minecraft\util\datafix\fixes\RidingToPassengers.java
+++ minecraft_patched\net\minecraft\util\datafix\fixes\RidingToPassengers.java
@@ -71,9 +71,19 @@
                      });
                   }
 
+                  try
+                  {
                   pair = (Pair)(new Typed(p_206340_3_, p_209760_5_, optional1.get())).getOptional(opticfinder).orElseThrow(() -> {
                      return new IllegalStateException("Should always have an entity here");
                   });
+                  }
+                  catch(Throwable e)
+                  {
+                    if(e instanceof RuntimeException)
+                      throw (RuntimeException)e;
+                    else
+                      throw new RuntimeException(e.getMessage(), e.getCause());
+                  }                  
                }
             };
          }), this.writeAndRead("player RootVehicle injecter", type4, type5));
