diff -r -U 3 minecraft\net\minecraft\util\math\shapes\IndirectMerger.java minecraft_patched\net\minecraft\util\math\shapes\IndirectMerger.java
--- minecraft\net\minecraft\util\math\shapes\IndirectMerger.java
+++ minecraft_patched\net\minecraft\util\math\shapes\IndirectMerger.java
@@ -35,7 +35,7 @@
          boolean flag2 = flag && (!flag1 || p_i47685_1_.getDouble(i) < p_i47685_2_.getDouble(j) + 1.0E-7D);
          double d1 = flag2 ? p_i47685_1_.getDouble(i++) : p_i47685_2_.getDouble(j++);
          if ((i != 0 && flag || flag2 || p_i47685_4_) && (j != 0 && flag1 || !flag2 || p_i47685_3_)) {
-            if (d0 <= d1 - 1.0E-7D) {
+           if (!(d0 > d1 - 1.0E-7D)) {
                this.field_197857_b.add(i - 1);
                this.field_197858_c.add(j - 1);
                this.field_197856_a.add(d1);
