diff -r -U 3 minecraft\net\minecraft\world\chunk\storage\AnvilChunkLoader.java minecraft_patched\net\minecraft\world\chunk\storage\AnvilChunkLoader.java
--- minecraft\net\minecraft\world\chunk\storage\AnvilChunkLoader.java
+++ minecraft_patched\net\minecraft\world\chunk\storage\AnvilChunkLoader.java
@@ -34,6 +34,7 @@
 import net.minecraft.nbt.NBTTagShort;
 import net.minecraft.nbt.NBTUtil;
 import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.registry.RegistryNamespacedDefaultedByKey;
@@ -164,11 +164,7 @@
       try {
          this.field_183014_e = true;
 
-         while(true) {
-            if (this.func_75814_c()) {
-               continue;
-            }
-         }
+         while(this.func_75814_c());
       } finally {
          this.field_183014_e = false;
       }
@@ -465,7 +465,7 @@
       }
 
       UpgradeData upgradedata = p_75823_2_.func_150297_b("UpgradeData", 10) ? new UpgradeData(p_75823_2_.func_74775_l("UpgradeData")) : UpgradeData.field_196994_a;
-      Predicate predicate = (p_205531_0_) -> {
+      Predicate<Block> predicate = (p_205531_0_) -> {
          return p_205531_0_.func_176223_P().func_196958_f();
       };
       IRegistry iregistry1 = IRegistry.field_501011_g;
@@ -473,8 +473,8 @@
       Function function = iregistry1::func_501993_b;
       IRegistry iregistry2 = IRegistry.field_501011_g;
       IRegistry.field_501011_g.getClass();
-      ChunkPrimerTickList<Block> chunkprimerticklist1 = new ChunkPrimerTickList<Block>(predicate, function, iregistry2::func_501994_a, new ChunkPos(i, j));
-      predicate = (p_206242_0_) -> {
+      ChunkPrimerTickList<Block> chunkprimerticklist1 = new ChunkPrimerTickList<Block>(predicate, iregistry1::func_501993_b, IRegistry.field_501011_g::func_501994_a, new ChunkPos(i, j));
+      Predicate<Fluid> predicateFluid = (p_206242_0_) -> {
          return p_206242_0_ == Fluids.field_204541_a;
       };
       IRegistry iregistry = IRegistry.field_501012_h;
@@ -482,7 +482,7 @@
       Function function1 = iregistry::func_501993_b;
       iregistry2 = IRegistry.field_501012_h;
       IRegistry.field_501012_h.getClass();
-      ChunkPrimerTickList<Fluid> chunkprimerticklist = new ChunkPrimerTickList<Fluid>(predicate, function1, iregistry2::func_501994_a, new ChunkPos(i, j));
+      ChunkPrimerTickList<Fluid> chunkprimerticklist = new ChunkPrimerTickList<Fluid>(predicateFluid, iregistry::func_501993_b, IRegistry.field_501012_h::func_501994_a, new ChunkPos(i, j));
       long l = p_75823_2_.func_74763_f("InhabitedTime");
       Chunk chunk = new Chunk(p_75823_1_.func_201672_e(), i, j, abiome, upgradedata, chunkprimerticklist1, chunkprimerticklist, l);
       chunk.func_201613_c(p_75823_2_.func_74779_i("Status"));
