--- a/net/minecraft/client/gui/toasts/SystemToast.java
+++ b/net/minecraft/client/gui/toasts/SystemToast.java
@@ -27,18 +27,18 @@
             this.newDisplay = false;
         }
 
-        toastGui.getMinecraft().getTextureManager().bindTexture(TEXTURE_TOASTS);
+        toastGui.getInstance().getTextureManager().bindTexture(TEXTURE_TOASTS);
         GlStateManager.color3f(1.0F, 1.0F, 1.0F);
         toastGui.blit(0, 0, 0, 64, 160, 32);
 
         if (this.subtitle == null)
         {
-            toastGui.getMinecraft().fontRenderer.drawString(this.title, 18.0F, 12.0F, -256);
+            toastGui.getInstance().fontRenderer.drawString(this.title, 18.0F, 12.0F, -256);
         }
         else
         {
-            toastGui.getMinecraft().fontRenderer.drawString(this.title, 18.0F, 7.0F, -256);
-            toastGui.getMinecraft().fontRenderer.drawString(this.subtitle, 18.0F, 18.0F, -1);
+            toastGui.getInstance().fontRenderer.drawString(this.title, 18.0F, 7.0F, -256);
+            toastGui.getInstance().fontRenderer.drawString(this.subtitle, 18.0F, 18.0F, -1);
         }
 
         return delta - this.firstDrawTime < 5000L ? IToast.Visibility.SHOW : IToast.Visibility.HIDE;
