--- a/net/minecraft/client/shader/Shader.java
+++ b/net/minecraft/client/shader/Shader.java
@@ -80,7 +80,8 @@
         this.manager.getShaderUniformOrDefault("OutSize").set(f, f1);
         this.manager.getShaderUniformOrDefault("Time").set(partialTicks);
         Minecraft minecraft = Minecraft.getMinecraft();
-        this.manager.getShaderUniformOrDefault("ScreenSize").set((float)minecraft.mainWindow.getFramebufferWidth(), (float)minecraft.mainWindow.getFramebufferHeight());
+        // VIVECRAFT: Use eye framebuffer instead of window
+        this.manager.getShaderUniformOrDefault("ScreenSize").set((float)minecraft.stereoProvider.framebufferVrRender.framebufferWidth, (float)minecraft.stereoProvider.framebufferVrRender.framebufferHeight);
         this.manager.useShader();
         this.framebufferOut.framebufferClear();
         this.framebufferOut.bindFramebuffer(false);
