--- a/net/minecraft/client/shader/Shader.java
+++ b/net/minecraft/client/shader/Shader.java
@@ -80,7 +80,8 @@
         this.manager.getShaderUniform("OutSize").set(f, f1);
         this.manager.getShaderUniform("Time").set(partialTicks);
         Minecraft minecraft = Minecraft.getInstance();
-        this.manager.getShaderUniform("ScreenSize").set((float)minecraft.mainWindow.getFramebufferWidth(), (float)minecraft.mainWindow.getFramebufferHeight());
+        // VIVECRAFT: Use eye framebuffer instead of window
+        this.manager.getShaderUniform("ScreenSize").set((float)minecraft.stereoProvider.framebufferVrRender.framebufferWidth, (float)minecraft.stereoProvider.framebufferVrRender.framebufferHeight);
         this.manager.func_216535_f();
         this.framebufferOut.framebufferClear(Minecraft.IS_RUNNING_ON_MAC);
         this.framebufferOut.bindFramebuffer(false);
