--- a/net/minecraft/entity/projectile/EntityTrident.java
+++ b/net/minecraft/entity/projectile/EntityTrident.java
@@ -1,6 +1,10 @@
 package net.minecraft.entity.projectile;
 
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.VivePlayer;
+
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
@@ -80,7 +84,17 @@
             else if (i > 0)
             {
                 this.func_203045_n(true);
-                Vec3d vec3d = new Vec3d(entity.posX - this.posX, entity.posY + (double)entity.getEyeHeight() - this.posY, entity.posZ - this.posZ);
+                
+                //Vivecraft       
+                Vec3d pos = new Vec3d(entity.posX, entity.posY + entity.getEyeHeight(), entity.posZ);
+            	VivePlayer vp = NetworkHelper.vivePlayers.get(entity.getUniqueID());
+            	if (vp !=null && vp.isVR())
+            	{
+            	        pos = vp.getControllerPos(0);
+            	}           	         	
+                Vec3d vec3d = new Vec3d(pos.x - this.posX, pos.y- this.posY, pos.z - this.posZ);
+                //
+                
                 this.posY += vec3d.y * 0.015D * (double)i;
 
                 if (this.world.isRemote)
