--- a/net/minecraft/entity/projectile/EntityFireball.java
+++ b/net/minecraft/entity/projectile/EntityFireball.java
@@ -1,5 +1,8 @@
 package net.minecraft.entity.projectile;
 
+import com.mtbs3d.minecrift.api.NetworkHelper;
+import com.mtbs3d.minecrift.api.VivePlayer;
+
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.EntityType;
@@ -40,11 +43,24 @@
         this.accelerationZ = p_i48544_12_ / d0 * 0.1D;
     }
 
-    public EntityFireball(EntityType<?> type, EntityLivingBase p_i48545_2_, double p_i48545_3_, double p_i48545_5_, double p_i48545_7_, World p_i48545_9_, float p_i48545_10_, float p_i48545_11_)
+    public EntityFireball(EntityType<?> type, EntityLivingBase shooter, double p_i48545_3_, double p_i48545_5_, double p_i48545_7_, World p_i48545_9_, float p_i48545_10_, float p_i48545_11_)
     {
         this(type, p_i48545_9_, p_i48545_10_, p_i48545_11_);
-        this.shootingEntity = p_i48545_2_;
-        this.setLocationAndAngles(p_i48545_2_.posX, p_i48545_2_.posY, p_i48545_2_.posZ, p_i48545_2_.rotationYaw, p_i48545_2_.rotationPitch);
+        this.shootingEntity = shooter;
+
+        //VIVE USE CONTROLLER
+    	VivePlayer vp = NetworkHelper.vivePlayers.get(shooter.getUniqueID());
+    	if (vp !=null && vp.isVR())
+    	{
+    		Vec3d aim = vp.getControllerDir(0);
+    		Vec3d aimSource = vp.getControllerPos(0);
+    		float pitch = (float)Math.toDegrees(Math.asin(aim.y/aim.length()));
+    		float yaw = (float)Math.toDegrees(Math.atan2(aim.x, -aim.z));   
+    		this.setLocationAndAngles(aimSource.x + aim.x*0.6f, aimSource.y+aim.y*0.6f, aimSource.z+aim.z*0.6f, yaw, pitch);
+        } else {
+        	this.setLocationAndAngles(shooter.posX, shooter.posY, shooter.posZ, shooter.rotationYaw, shooter.rotationPitch);
+        }
+
         this.setPosition(this.posX, this.posY, this.posZ);
         this.motionX = 0.0D;
         this.motionY = 0.0D;
@@ -56,6 +72,7 @@
         this.accelerationX = p_i48545_3_ / d0 * 0.1D;
         this.accelerationY = p_i48545_5_ / d0 * 0.1D;
         this.accelerationZ = p_i48545_7_ / d0 * 0.1D;
+
     }
 
     protected void registerData()
