--- a/net/minecraft/client/particle/ParticleManager.java
+++ b/net/minecraft/client/particle/ParticleManager.java
@@ -193,7 +193,7 @@
                     Stream stream = p_lambda$null$4_3_.stream();
                     AtlasTexture atlastexture = this.atlas;
                     this.atlas.getClass();
-                    immutablelist = stream.map(atlastexture::getSprite).collect(ImmutableList.toImmutableList());
+                    immutablelist = (ImmutableList) ((Stream<ResourceLocation>)stream).map(atlastexture::getSprite).collect(ImmutableList.toImmutableList());
                 }
 
                 this.sprites.get(p_lambda$null$4_2_).func_217592_a(immutablelist);
