--- a/net/minecraft/client/particle/ParticleManager.java
+++ b/net/minecraft/client/particle/ParticleManager.java
@@ -173,7 +173,7 @@
             preparationsProfiler.endSection();
             preparationsProfiler.endTick();
             return atlastexture$sheetdata;
-        }, backgroundExecutor).<Void>thenCompose(stage::markCompleteAwaitingOthers).thenAcceptAsync((p_lambda$reload$5_3_) ->
+        }, backgroundExecutor).thenCompose(stage::markCompleteAwaitingOthers).thenAcceptAsync((p_lambda$reload$5_3_) ->
         {
             reloadProfiler.startTick();
             reloadProfiler.startSection("upload");
