--- a/net/minecraft/client/gui/screen/inventory/ContainerScreen.java
+++ b/net/minecraft/client/gui/screen/inventory/ContainerScreen.java
@@ -110,9 +110,9 @@
                 GlStateManager.disableDepthTest();
                 int j1 = slot.xPos;
                 int k1 = slot.yPos;
-                GlStateManager.colorMask(true, true, true, false);
+                //GlStateManager.colorMask(true, true, true, false);
                 this.fillGradient(j1, k1, j1 + 16, k1 + 16, -2130706433, -2130706433);
-                GlStateManager.colorMask(true, true, true, true);
+                //GlStateManager.colorMask(true, true, true, true);
                 GlStateManager.enableLighting();
                 GlStateManager.enableDepthTest();
             }
@@ -397,7 +397,8 @@
                             }
                             else
                             {
-                                boolean flag2 = l != -999 && (InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 344));
+                            	//Vivecraft add shift
+                                boolean flag2 = l != -999 && (isFakeShift() || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 344));
                                 ClickType clicktype = ClickType.PICKUP;
 
                                 if (flag2)
@@ -617,7 +618,8 @@
                 }
                 else
                 {
-                    boolean flag1 = k != -999 && (InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 344));
+                	//Vivecraft add shift
+                    boolean flag1 = k != -999 && (isFakeShift() || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 344));
 
                     if (flag1)
                     {
