--- a/net/minecraft/client/gui/advancements/GuiAdvancement.java
+++ b/net/minecraft/client/gui/advancements/GuiAdvancement.java
@@ -180,7 +180,11 @@
             this.minecraft.getTextureManager().bindTexture(WIDGETS);
             GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
             GlStateManager.enableBlend();
+            //VIVECRAFT Prevent invisible background - test if needed in 1.13
+            GlStateManager.colorMask(true, true, true, false);
             this.drawTexturedModalRect(p_191817_1_ + this.x + 3, p_191817_2_ + this.y, this.displayInfo.getFrame().getIcon(), 128 + advancementstate.getId() * 26, 26, 26);
+            GlStateManager.colorMask(true, true, true, true);
+            //
             RenderHelper.enableGUIStandardItemLighting();
             this.minecraft.getItemRenderer().renderItemAndEffectIntoGUI((EntityLivingBase)null, this.displayInfo.getIcon(), p_191817_1_ + this.x + 8, p_191817_2_ + this.y + 5);
         }
