--- a/net/minecraft/client/renderer/entity/EntityRendererManager.java
+++ b/net/minecraft/client/renderer/entity/EntityRendererManager.java
@@ -131,7 +131,7 @@
 
 public class EntityRendererManager
 {
-    private final Map < Class <? extends Entity > , EntityRenderer <? extends Entity >> renderers = Maps.newHashMap();
+    private final Map <Class, EntityRenderer> renderers = Maps.newHashMap();
     private final Map<String, PlayerRenderer> skinMap = Maps.newHashMap();
     private final PlayerRenderer playerRenderer;
     private FontRenderer textRenderer;
