--- a/net/minecraft/client/world/ClientWorld.java
+++ b/net/minecraft/client/world/ClientWorld.java
@@ -565,7 +565,7 @@
 
     public void playSound(double x, double y, double z, SoundEvent soundIn, SoundCategory category, float volume, float pitch, boolean distanceDelay)
     {
-        double d0 = this.mc.gameRenderer.getActiveRenderInfo().getProjectedView().squareDistanceTo(x, y, z);
+        double d0 = this.mc.gameRenderer.getActiveRenderInfo().getPosition().squareDistanceTo(x, y, z);
         SimpleSound simplesound = new SimpleSound(soundIn, category, volume, pitch, (float)x, (float)y, (float)z);
 
         if (distanceDelay && d0 > 100.0D)
