--- a/net/minecraft/client/MainWindow.java
+++ b/net/minecraft/client/MainWindow.java
@@ -375,7 +375,10 @@
 
     public void updateVsyncFromGameSettings()
     {
-        GLFW.glfwSwapInterval(this.mc.gameSettings.enableVsync ? 1 : 0);
+    	//Vivecraft
+    	GLFW.glfwSwapInterval(0);
+        //GLFW.glfwSwapInterval(this.mc.gameSettings.enableVsync ? 1 : 0);
+    	//
     }
 
     public void close()
@@ -597,16 +600,21 @@
             this.mc.currentScreen.onResize(this.mc, this.scaledWidth, this.scaledHeight);
         }
 
+        if (mc.stereoProvider != null)
+        	mc.stereoProvider.reinitFrameBuffers("Main Window Changed");
+        
         Framebuffer framebuffer = this.mc.getFramebuffer();
 
         if (framebuffer != null)
         {
-            framebuffer.createBindFramebuffer(this.framebufferWidth, this.framebufferHeight);
+        	//no plz.
+            //framebuffer.createBindFramebuffer(this.framebufferWidth, this.framebufferHeight);
         }
 
         if (this.mc.entityRenderer != null)
         {
-            this.mc.entityRenderer.updateShaderGroupSize(this.framebufferWidth, this.framebufferHeight);
+            // VIVECRAFT: Use eye framebuffer instead of window
+            this.mc.entityRenderer.updateShaderGroupSize(this.mc.stereoProvider.framebufferVrRender.framebufferTextureWidth, this.mc.stereoProvider.framebufferVrRender.framebufferTextureHeight);
         }
 
         if (this.mc.mouseHelper != null)
