--- a/net/minecraft/client/MainWindow.java
+++ b/net/minecraft/client/MainWindow.java
@@ -124,9 +124,9 @@
         }
     }
 
-    public void func_216522_a(boolean p_216522_1_)
+    public void setupOverlayRendering(boolean is_mac)
     {
-        GlStateManager.clear(256, p_216522_1_);
+        GlStateManager.clear(256, is_mac);
         GlStateManager.matrixMode(5889);
         GlStateManager.loadIdentity();
         GlStateManager.ortho(0.0D, (double)this.getFramebufferWidth() / this.getGuiScaleFactor(), (double)this.getFramebufferHeight() / this.getGuiScaleFactor(), 0.0D, 1000.0D, 3000.0D);
