--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -186,6 +186,7 @@
 import net.minecraft.util.registry.Registry;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.KeybindTextComponent;
+import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.util.text.TextFormatting;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.Difficulty;
@@ -200,6 +201,7 @@
 import net.minecraft.world.storage.WorldInfo;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.render.PlayerModelController;
 
 public class Minecraft extends RecursiveEventLoop<Runnable> implements ISnooperInfo, IWindowEventListener, AutoCloseable
 {
@@ -461,7 +463,7 @@
         }
 
         this.virtualScreen = new VirtualScreen(this);
-        this.mainWindow = this.virtualScreen.create(screensize, this.gameSettings.fullscreenResolution, "Minecraft " + SharedConstants.getVersion().getName());
+        this.mainWindow = this.virtualScreen.create(screensize, this.gameSettings.fullscreenResolution, this.minecriftVerString);
         this.setGameFocused(true);
 
         try
@@ -1563,7 +1565,11 @@
         this.gameRenderer.getMouseOver(1.0F);
         this.tutorial.onMouseHover(this.world, this.objectMouseOver);
         this.profiler.startSection("gameMode");
-
+        
+        //Vivecraft
+       	PlayerModelController.getInstance().tick();
+       	//
+        
         if (!this.isGamePaused && this.world != null)
         {
             this.playerController.tick();
@@ -2769,4 +2775,16 @@
     {
         return this.loadingGui;
     }
+    
+    //Vivecraft Additions
+
+    public final String minecriftVerString = "Vivecraft 1.14.4 jrbudda-NONVR-3-a1";
+
+    public void printChatMessage(String msg)
+    {
+    	if (this.world != null) {
+    		ITextComponent chatText = new StringTextComponent(msg);
+    		this.ingameGUI.getChatGUI().printChatMessage(chatText);
+    	}
+    }
 }
