--- a/net/minecraft/client/renderer/entity/layers/MooshroomMushroomLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/MooshroomMushroomLayer.java
@@ -27,8 +27,8 @@
     public MooshroomMushroomLayer(IEntityRenderer<T, CowModel<T>> p_i14_1_)
     {
         super(p_i14_1_);
-        MooshroomRenderer mooshroomrenderer = (MooshroomRenderer)p_i14_1_;
-        this.modelRendererMushroom = new RendererModel(mooshroomrenderer.entityModel);
+        IEntityRenderer<T, CowModel<T>>  mooshroomrenderer = p_i14_1_;
+        this.modelRendererMushroom = new RendererModel(mooshroomrenderer.getEntityModel());
         this.modelRendererMushroom.setTextureSize(16, 16);
         this.modelRendererMushroom.rotationPointX = -6.0F;
         this.modelRendererMushroom.rotationPointZ = -8.0F;
