--- a/net/optifine/shaders/ShadersRender.java
+++ b/net/optifine/shaders/ShadersRender.java
@@ -2,6 +2,7 @@
 
 import java.nio.IntBuffer;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.ActiveRenderInfo;
 import net.minecraft.client.renderer.BufferBuilder;
 import net.minecraft.client.renderer.EntityRenderer;
 import net.minecraft.client.renderer.FirstPersonRenderer;
@@ -110,7 +111,7 @@
             {
                 Shaders.readCenterDepth();
                 Shaders.beginHand(false);
-                GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+                GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
                 Shaders.setSkipRenderHands(flag, flag1);
                 er.renderHand(partialTicks, true, false, false);
                 Shaders.endHand();
@@ -127,7 +128,7 @@
             Shaders.readCenterDepth();
             GlStateManager.enableBlend();
             Shaders.beginHand(true);
-            GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+            GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
             Shaders.setSkipRenderHands(Shaders.isHandRenderedMain(), Shaders.isHandRenderedOff());
             er.renderHand(partialTicks, true, false, true);
             Shaders.endHand();
@@ -155,7 +156,7 @@
         }
 
         GlStateManager.depthFunc(515);
-        GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+        GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
         itemRenderer.renderItemInFirstPerson(par1);
         Shaders.setRenderingFirstPersonHand(false);
     }
@@ -202,7 +203,7 @@
             RenderGlobal renderglobal = minecraft.renderGlobal;
             Shaders.isShadowPass = true;
             Shaders.shadowPassCounter = Shaders.shadowPassInterval;
-            Shaders.preShadowPassThirdPersonView = minecraft.gameSettings.thirdPersonView;
+            Shaders.preShadowPassThirdPersonView =0; //Vivecraft minecraft.gameSettings.thirdPersonView;
             minecraft.gameSettings.thirdPersonView = 1;
             Shaders.checkGLError("pre shadow");
             GL11.glMatrixMode(GL11.GL_PROJECTION);
@@ -214,7 +215,18 @@
             Shaders.checkGLError("shadow bind sfb");
             minecraft.profiler.endStartSection("shadow camera");
             entityRenderer.setupCameraTransform(partialTicks);
+            
+            //vivecraft
+            // I don't think this actually has any effect...
+            entityRenderer.applyCameraDepth(false);
+            //
+            
             Shaders.setCameraShadow(partialTicks);
+            
+            //Vivecraft
+            ActiveRenderInfo.updateRenderInfo(minecraft.player, minecraft.gameSettings.thirdPersonView == 2, minecraft.entityRenderer.getFarPlaneDistance());
+            //
+
             Shaders.checkGLError("shadow camera");
             Shaders.useProgram(Shaders.ProgramShadow);
             GL20.glDrawBuffers(Shaders.sfbDrawBuffers);
@@ -229,7 +241,7 @@
             }
 
             Shaders.checkFramebufferStatus("shadow fb");
-            GL11.glClearColor(1.0F, 1.0F, 1.0F, 1.0F);
+            GlStateManager.clearColor(1.0F, 1.0F, 1.0F, 1.0F);
             GL11.glClear(Shaders.usedShadowColorBuffers != 0 ? GL11.GL_COLOR_BUFFER_BIT | GL11.GL_DEPTH_BUFFER_BIT : GL11.GL_DEPTH_BUFFER_BIT);
             Shaders.checkGLError("shadow clear");
             minecraft.profiler.endStartSection("shadow frustum");
@@ -262,6 +274,9 @@
             renderglobal.renderBlockLayer(BlockRenderLayer.SOLID, (double)partialTicks, entity);
             Shaders.checkGLError("shadow terrain solid");
             GlStateManager.enableAlphaTest();
+            //Vivecraft
+            GlStateManager.alphaFunc(516, 0.1F);
+            //
             renderglobal.renderBlockLayer(BlockRenderLayer.CUTOUT_MIPPED, (double)partialTicks, entity);
             Shaders.checkGLError("shadow terrain cutoutmipped");
             minecraft.getTextureManager().getTexture(TextureMap.LOCATION_BLOCKS_TEXTURE).setBlurMipmap(false, false);
@@ -329,7 +344,8 @@
             GlStateManager.depthMask(true);
             GlStateManager.enableCull();
             GlStateManager.disableBlend();
-            GL11.glFlush();
+            // VIVECRAFT: yeah no
+            //GL11.glFlush();
             Shaders.checkGLError("shadow flush");
             Shaders.isShadowPass = false;
             minecraft.gameSettings.thirdPersonView = Shaders.preShadowPassThirdPersonView;
