--- a/net/optifine/shaders/ShadersRender.java
+++ b/net/optifine/shaders/ShadersRender.java
@@ -144,14 +144,14 @@
         if (renderTranslucent)
         {
             GlStateManager.depthFunc(519);
-            GL11.glPushMatrix();
+            GlStateManager.pushMatrix();
             IntBuffer intbuffer = Shaders.activeDrawBuffers;
             Shaders.setDrawBuffers(Shaders.drawBuffersNone);
             Shaders.renderItemKeepDepthMask = true;
             itemRenderer.renderItemInFirstPerson(par1);
             Shaders.renderItemKeepDepthMask = false;
             Shaders.setDrawBuffers(intbuffer);
-            GL11.glPopMatrix();
+            GlStateManager.popMatrix();
         }
 
         GlStateManager.depthFunc(515);
@@ -193,7 +193,7 @@
         }
     }
 
-    public static void renderShadowMap(GameRenderer entityRenderer, ActiveRenderInfo activeRenderInfo, int pass, float partialTicks, long finishTimeNano)
+    public static void renderShadowMap(GameRenderer gameRenderer, ActiveRenderInfo activeRenderInfo, int pass, float partialTicks, long finishTimeNano)
     {
         if (Shaders.usedShadowDepthBuffers > 0 && --Shaders.shadowPassCounter <= 0)
         {
@@ -204,14 +204,20 @@
             Shaders.shadowPassCounter = Shaders.shadowPassInterval;
             Shaders.checkGLError("pre shadow");
             GL30.glMatrixMode(5889);
-            GL11.glPushMatrix();
+            GlStateManager.pushMatrix();
             GL30.glMatrixMode(5888);
-            GL11.glPushMatrix();
+            GlStateManager.pushMatrix();
             minecraft.getProfiler().endStartSection("shadow clear");
             EXTFramebufferObject.glBindFramebufferEXT(36160, Shaders.sfb);
             Shaders.checkGLError("shadow bind sfb");
             minecraft.getProfiler().endStartSection("shadow camera");
-            entityRenderer.setupCameraTransform(partialTicks);
+            gameRenderer.setupCameraTransform(partialTicks);
+
+             //vivecraft
+            // I don't think this actually has any effect...
+            gameRenderer.applyCameraDepth(false);
+            //  
+                      
             updateActiveRenderInfo(activeRenderInfo, minecraft, partialTicks);
             Shaders.setCameraShadow(activeRenderInfo, partialTicks);
             Shaders.checkGLError("shadow camera");
@@ -247,8 +253,8 @@
             minecraft.getTextureManager().bindTexture(AtlasTexture.LOCATION_BLOCKS_TEXTURE);
             minecraft.getProfiler().endStartSection("shadow setupterrain");
             int i = 0;
-            i = entityRenderer.frameCount;
-            entityRenderer.frameCount = i + 1;
+            i = gameRenderer.frameCount;
+            gameRenderer.frameCount = i + 1;
             worldrenderer.setupTerrain(activeRenderInfo, frustum, i, minecraft.player.isSpectator());
             minecraft.getProfiler().endStartSection("shadow updatechunks");
             minecraft.getProfiler().endStartSection("shadow terrain");
@@ -258,6 +264,9 @@
             worldrenderer.renderBlockLayer(BlockRenderLayer.SOLID, activeRenderInfo);
             Shaders.checkGLError("shadow terrain solid");
             GlStateManager.enableAlphaTest();
+            //Vivecraft
+            GlStateManager.alphaFunc(516, 0.1F);
+            //
             worldrenderer.renderBlockLayer(BlockRenderLayer.CUTOUT_MIPPED, activeRenderInfo);
             Shaders.checkGLError("shadow terrain cutoutmipped");
             minecraft.getTextureManager().getTexture(AtlasTexture.LOCATION_BLOCKS_TEXTURE).setBlurMipmap(false, false);
@@ -309,7 +318,8 @@
             GlStateManager.depthMask(true);
             GlStateManager.enableCull();
             GlStateManager.disableBlend();
-            GL30.glFlush();
+            // VIVECRAFT: yeah still no 3.0
+            //GL30.glFlush();
             Shaders.checkGLError("shadow flush");
             Shaders.isShadowPass = false;
             minecraft.getProfiler().endStartSection("shadow postprocess");
@@ -366,9 +376,9 @@
             minecraft.getTextureManager().bindTexture(AtlasTexture.LOCATION_BLOCKS_TEXTURE);
             Shaders.useProgram(Shaders.ProgramTerrain);
             GL30.glMatrixMode(5888);
-            GL11.glPopMatrix();
+            GlStateManager.popMatrix();
             GL30.glMatrixMode(5889);
-            GL11.glPopMatrix();
+            GlStateManager.popMatrix();
             GL30.glMatrixMode(5888);
             Shaders.checkGLError("shadow end");
         }
