--- a/net/minecraft/client/renderer/texture/AtlasTexture.java
+++ b/net/minecraft/client/renderer/texture/AtlasTexture.java
@@ -208,7 +208,7 @@
                 {
                     GlStateManager.bindTexture(this.getMultiTexID().norm);
 
-                    for (TextureAtlasSprite textureatlassprite2 : list)
+                    for (TextureAtlasSprite textureatlassprite2 : (Set<TextureAtlasSprite>)(Set<?>)list)
                     {
                         TextureAtlasSprite textureatlassprite4 = textureatlassprite2.spriteNormal;
 
@@ -223,7 +223,7 @@
                 {
                     GlStateManager.bindTexture(this.getMultiTexID().spec);
 
-                    for (TextureAtlasSprite textureatlassprite3 : list)
+                    for (TextureAtlasSprite textureatlassprite3 : (Set<TextureAtlasSprite>)(Set<?>)list)
                     {
                         TextureAtlasSprite textureatlassprite5 = textureatlassprite3.spriteSpecular;
 
@@ -776,7 +776,7 @@
                 if (!textureatlassprite.hasCustomLoader(p_loadTexture_2_, p_loadTexture_3_) || !textureatlassprite.load(p_loadTexture_2_, p_loadTexture_3_, this.mapUploadedSprites::get))
                 {
                     p_loadTexture_4_ = Math.min(p_loadTexture_4_, Math.min(textureatlassprite.getWidth(), textureatlassprite.getHeight()));
-                    j1 = Math.min(Integer.lowestOneBit(textureatlassprite.getWidth()), Integer.lowestOneBit(textureatlassprite.getHeight()));
+                    i = Math.min(Integer.lowestOneBit(textureatlassprite.getWidth()), Integer.lowestOneBit(textureatlassprite.getHeight()));
 
                     if (i < p_loadTexture_5_)
                     {
@@ -924,7 +924,7 @@
                             }
                             else
                             {
-                                int k = map.get(j);
+                                int k = (int) map.get(j);
                                 map.put(j, k + 1);
                             }
                         }
@@ -938,12 +938,12 @@
         }
 
         int l = 0;
-        Set set = map.keySet();
-        Set set1 = new TreeSet(set);
+        Set<Integer> set = map.keySet();
+        Set<Integer> set1 = new TreeSet(set);
 
         for (int j1 : set1)
         {
-            int l1 = map.get(j1);
+            int l1 = (int) map.get(j1);
             l += l1;
         }
 
@@ -953,7 +953,7 @@
 
         for (int j2 : set1)
         {
-            int k2 = map.get(j2);
+            int k2 = (int) map.get(j2);
             k1 += k2;
 
             if (j2 > i1)
