--- a/net/minecraft/client/renderer/texture/DownloadingTexture.java
+++ b/net/minecraft/client/renderer/texture/DownloadingTexture.java
@@ -269,10 +269,10 @@
             {
                 nativeimage = this.imageBuffer.parseUserSkin(nativeimage);
             }
-
+            final NativeImage nif = nativeimage; 
             Minecraft.getInstance().execute(() ->
             {
-                this.setImage(nativeimage);
+                this.setImage(nif);
             });
             this.uploadPending = true;
         }
