--- a/net/minecraft/client/renderer/debug/DebugRenderer.java
+++ b/net/minecraft/client/renderer/debug/DebugRenderer.java
@@ -5,6 +5,7 @@
 import net.minecraft.client.renderer.GlStateManager;
 import net.minecraft.client.renderer.entity.RenderManager;
 import net.minecraft.entity.player.EntityPlayer;
+import org.vivecraft.utils.Debug;
 
 public class DebugRenderer
 {
@@ -48,7 +49,7 @@
 
     public boolean shouldRender()
     {
-        return this.chunkBorderEnabled || this.pathfindingEnabled || this.waterEnabled || this.heightMapEnabled || this.collisionBoxEnabled || this.neighborsUpdateEnabled || this.lightEnabled || this.worldGenAttemptsEnabled || this.solidFaceEnabled;
+        return this.chunkBorderEnabled || this.pathfindingEnabled || this.waterEnabled || this.heightMapEnabled || this.collisionBoxEnabled || this.neighborsUpdateEnabled || this.lightEnabled || this.worldGenAttemptsEnabled || this.solidFaceEnabled || Debug.isEnabled;
     }
 
     /**
@@ -116,6 +117,10 @@
         {
             this.solidFace.render(partialTicks, finishTimeNano);
         }
+	
+		if(Debug.isEnabled){
+			Debug.getRenderer().render(partialTicks,finishTimeNano);
+		}
     }
 
     public static void renderDebugText(String str, int x, int y, int z, float partialTicks, int color)
