--- a/net/minecraft/src/Config.java
+++ b/net/minecraft/src/Config.java
@@ -132,13 +132,14 @@
 
     public static void initGameSettings(GameSettings p_initGameSettings_0_)
     {
-        if (gameSettings == null)
-        {
+        //VIVE WHY WOULD YOU DO THAT??
+        //if (gameSettings == null)
+        //{
             gameSettings = p_initGameSettings_0_;
             updateAvailableProcessors();
             ReflectorForge.putLaunchBlackboard("optifine.ForgeSplashCompatible", Boolean.TRUE);
             antialiasingLevel = gameSettings.ofAaLevel;
-        }
+        //}
     }
 
     public static void initDisplay()
