--- a/net/minecraft/client/renderer/entity/model/ModelLargeChest.java
+++ b/net/minecraft/client/renderer/entity/model/ModelLargeChest.java
@@ -15,9 +15,70 @@
         this.chestKnob.rotationPointY = 7.0F;
         this.chestKnob.rotationPointZ = 15.0F;
         this.chestBelow = (new ModelRenderer(this, 0, 19)).setTextureSize(128, 64);
-        this.chestBelow.addBox(0.0F, 0.0F, 0.0F, 30, 10, 14, 0.0F);
+		makeHole();
+        //this.chestBelow.addBox(0.0F, 0.0F, 0.0F, 30, 10, 14, 0.0F);
         this.chestBelow.rotationPointX = 1.0F;
         this.chestBelow.rotationPointY = 6.0F;
         this.chestBelow.rotationPointZ = 1.0F;
     }
+	
+	@Override
+	void makeHole(){
+		
+		int[][] texturemap=new int[6][];
+		texturemap[0]=new int[]{44,19,74,33};//bottom
+		texturemap[1]=new int[]{0,0,0,0};//top
+		texturemap[2]=new int[]{14,33,44,43};//front
+		texturemap[3]=new int[]{58,33,88,43};//back
+		texturemap[4]=new int[]{0,33,14,43};//left
+		texturemap[5]=new int[]{44,33,58,43};//right
+		
+		int[][] texturemap2=new int[6][];
+		texturemap2[0]=new int[]{0,0,0,0};//bottom
+		texturemap2[1]=new int[]{14,19,16,33};//top
+		texturemap2[2]=new int[]{0,0,0,0};//front
+		texturemap2[3]=new int[]{0,0,0,0};//back
+		texturemap2[4]=new int[]{0,0,0,0};//left
+		texturemap2[5]=new int[]{44,1,58,11};//right
+		
+		int[][] texturemap3=new int[6][];
+		texturemap3[0]=new int[]{0,0,0,0};//bottom
+		texturemap3[1]=new int[]{42,19,44,33};//top
+		texturemap3[2]=new int[]{0,0,0,0};//front
+		texturemap3[3]=new int[]{0,0,0,0};//back
+		texturemap3[4]=new int[]{44,1,58,11};//left
+		texturemap3[5]=new int[]{0,0,0,0};//right
+		
+		int[][] texturemap4=new int[6][];
+		texturemap4[0]=new int[]{0,0,0,0};//bottom
+		texturemap4[1]=new int[]{16,19,42,21};//top
+		texturemap4[2]=new int[]{0,0,0,0};//front
+		texturemap4[3]=new int[]{48,1,72,11};//back
+		texturemap4[4]=new int[]{0,0,0,0};//left
+		texturemap4[5]=new int[]{0,0,0,0};//right
+		
+		int[][] texturemap5=new int[6][];
+		texturemap5[0]=new int[]{0,0,0,0};//bottom
+		texturemap5[1]=new int[]{16,31,42,33};//top
+		texturemap5[2]=new int[]{48,1,72,11};//front
+		texturemap5[3]=new int[]{0,0,0,0};//back
+		texturemap5[4]=new int[]{0,0,0,0};//left
+		texturemap5[5]=new int[]{0,0,0,0};//right
+		
+		int[][] texturemap6=new int[6][];
+		texturemap6[0]=new int[]{0,0,0,0};//bottom
+		texturemap6[1]=new int[]{44,0,74,14};//top
+		texturemap6[2]=new int[]{0,0,0,0};//front
+		texturemap6[3]=new int[]{0,0,0,0};//back
+		texturemap6[4]=new int[]{0,0,0,0};//left
+		texturemap6[5]=new int[]{0,0,0,0};//right
+		
+		
+		this.chestBelow.cubeList.add(new ModelBox(chestBelow,texturemap,0,0,0,30,10,14,0,true));
+		this.chestBelow.cubeList.add(new ModelBox(chestBelow,texturemap2,0,0,0,2,10,14,0,true));
+		this.chestBelow.cubeList.add(new ModelBox(chestBelow,texturemap3,28,0,0,2,10,14,0,true));
+		this.chestBelow.cubeList.add(new ModelBox(chestBelow,texturemap4,2,0,0,26,10,2,0,true));
+		this.chestBelow.cubeList.add(new ModelBox(chestBelow,texturemap5,2,0,12,26,10,2,0,true));
+		this.chestBelow.cubeList.add(new ModelBox(chestBelow,texturemap6,0,8,0,30,2,14,0,true));
+	}
 }
