--- a/net/minecraft/client/renderer/tileentity/ChestTileEntityRenderer.java
+++ b/net/minecraft/client/renderer/tileentity/ChestTileEntityRenderer.java
@@ -2,9 +2,14 @@
 
 import com.mojang.blaze3d.platform.GlStateManager;
 import java.util.Calendar;
+
+import org.vivecraft.gui.physical.PhysicalChest;
+import org.vivecraft.gui.physical.PhysicalGui;
+
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
 import net.minecraft.block.ChestBlock;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.tileentity.model.ChestModel;
 import net.minecraft.client.renderer.tileentity.model.LargeChestModel;
 import net.minecraft.state.properties.ChestType;
@@ -123,9 +128,24 @@
 
     private void applyLidRotation(T p_199346_1_, float p_199346_2_, ChestModel p_199346_3_)
     {
-        float f = ((IChestLid)p_199346_1_).getLidAngle(p_199346_2_);
-        f = 1.0F - f;
-        f = 1.0F - f * f * f;
-        p_199346_3_.getLid().rotateAngleX = -(f * ((float)Math.PI / 2F));
+    	float f = ((IChestLid)p_199346_1_).getLidAngle(p_199346_2_);
+    	f = 1.0F - f;
+    	f = 1.0F - f * f * f;
+    	//VIVECRAFT START
+    	double overrideAngle=-1;
+    	for(PhysicalGui gui: Minecraft.getInstance().physicalGuiManager.guisInRange){
+    		if(gui instanceof PhysicalChest) {
+    			double tAngle = ((PhysicalChest) gui).getLidHoldAngle(p_199346_1_.getPos(), p_199346_2_);
+    			if(tAngle!=-1)
+    				overrideAngle=tAngle;
+    		}
+    	}
+
+    	if(overrideAngle!=-1){
+    		p_199346_3_.getLid().rotateAngleX=(float)Math.toRadians(-overrideAngle);
+    	}else {
+    		p_199346_3_.getLid().rotateAngleX = -(f * ((float)Math.PI / 2F));
+    	}
+    	//VIVECRAFT END    
     }
 }
