--- a/net/minecraft/client/renderer/tileentity/TileEntityRenderer.java
+++ b/net/minecraft/client/renderer/tileentity/TileEntityRenderer.java
@@ -88,7 +88,7 @@
     protected void drawNameplate(T te, String str, double x, double y, double z, int maxDistance)
     {
         ActiveRenderInfo activerenderinfo = this.rendererDispatcher.renderInfo;
-        double d0 = te.getDistanceSq(activerenderinfo.getProjectedView().x, activerenderinfo.getProjectedView().y, activerenderinfo.getProjectedView().z);
+        double d0 = te.getDistanceSq(activerenderinfo.getPosition().x, activerenderinfo.getPosition().y, activerenderinfo.getPosition().z);
 
         if (d0 <= (double)(maxDistance * maxDistance))
         {
