--- a/net/minecraft/client/renderer/tileentity/TileEntityChestRenderer.java
+++ b/net/minecraft/client/renderer/tileentity/TileEntityChestRenderer.java
@@ -3,6 +3,7 @@
 import java.util.Calendar;
 import net.minecraft.block.BlockChest;
 import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.GlStateManager;
 import net.minecraft.client.renderer.entity.model.ModelChest;
 import net.minecraft.client.renderer.entity.model.ModelLargeChest;
@@ -14,6 +15,8 @@
 import net.minecraft.tileentity.TileEntityTrappedChest;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.ResourceLocation;
+import org.vivecraft.gui.physical.PhysicalChest;
+import org.vivecraft.gui.physical.PhysicalGui;
 
 public class TileEntityChestRenderer<T extends TileEntity & IChestLid> extends TileEntityRenderer<T>
 {
@@ -121,11 +124,29 @@
         return p_199347_3_ ? this.largeChest : this.simpleChest;
     }
 
-    private void func_199346_a(T p_199346_1_, float p_199346_2_, ModelChest p_199346_3_)
+    private void func_199346_a(T tileEntity, float partialTicks, ModelChest modelchest)
     {
-        float f = ((IChestLid)p_199346_1_).getLidAngle(p_199346_2_);
+        float f = ((IChestLid)tileEntity).getLidAngle(partialTicks);
         f = 1.0F - f;
         f = 1.0F - f * f * f;
-        p_199346_3_.getLid().rotateAngleX = -(f * ((float)Math.PI / 2F));
+	
+		//VIVECRAFT START
+		double overrideAngle=-1;
+		for(PhysicalGui gui: Minecraft.getMinecraft().physicalGuiManager.guisInRange){
+			if(gui instanceof PhysicalChest) {
+				double tAngle = ((PhysicalChest) gui).getLidHoldAngle(tileEntity.getPos(), partialTicks);
+				if(tAngle!=-1)
+					overrideAngle=tAngle;
+			}
+		}
+	
+		if(overrideAngle!=-1){
+			modelchest.getLid().rotateAngleX=(float)Math.toRadians(-overrideAngle);
+		}else {
+			modelchest.getLid().rotateAngleX = -(f * ((float)Math.PI / 2F));
+		}
+		//VIVECRAFT END
+        
+        
     }
 }
