--- a/net/minecraft/client/renderer/block/model/ModelBakery.java
+++ b/net/minecraft/client/renderer/block/model/ModelBakery.java
@@ -89,7 +89,8 @@
     private final IResourceManager resourceManager;
     private final TextureMap textureMap;
     private final Map<ModelResourceLocation, IBakedModel> bakedRegistry = Maps.<ModelResourceLocation, IBakedModel>newHashMap();
-    private static final Map<ResourceLocation, StateContainer<Block, IBlockState>> field_209607_C = ImmutableMap.<ResourceLocation, StateContainer<Block, IBlockState>>of(new ResourceLocation("item_frame"), STATE_CONTAINER_ITEM_FRAME);
+    public static Map<ResourceLocation, StateContainer<Block, IBlockState>> field_209607_C = Maps.newHashMap();
+    static { field_209607_C.put(new ResourceLocation("item_frame"),STATE_CONTAINER_ITEM_FRAME); }
     private final Map<ResourceLocation, IUnbakedModel> unbakedModels = Maps.<ResourceLocation, IUnbakedModel>newHashMap();
     private final Set<ResourceLocation> unbakedModelLoadingQueue = Sets.<ResourceLocation>newHashSet();
     private final ModelBlockDefinition.ContainerHolder containerHolder = new ModelBlockDefinition.ContainerHolder();
