--- a/net/minecraft/src/Config.java
+++ b/net/minecraft/src/Config.java
@@ -1150,7 +1150,7 @@
             }
         }
 
-        IResourcePack[] airesourcepack = list.toArray(new IResourcePack[list.size()]);
+        IResourcePack[] airesourcepack = (IResourcePack[]) list.toArray(new IResourcePack[list.size()]);
         return airesourcepack;
     }
 
@@ -1471,7 +1471,7 @@
             list.add(s);
         }
 
-        String[] astring = list.toArray(new String[list.size()]);
+        String[] astring = (String[]) list.toArray(new String[list.size()]);
         return astring;
     }
 
@@ -1603,7 +1603,7 @@
 
             if (s == null)
             {
-                String[] astring = list.toArray(new String[list.size()]);
+                String[] astring = (String[]) list.toArray(new String[list.size()]);
                 return astring;
             }
 
@@ -1911,7 +1911,7 @@
         {
             int i = p_addObjectToArray_0_.length;
             int j = i + 1;
-            Object[] aobject = Array.newInstance(p_addObjectToArray_0_.getClass().getComponentType(), j);
+            Object[] aobject = (Object[]) Array.newInstance(p_addObjectToArray_0_.getClass().getComponentType(), j);
             System.arraycopy(p_addObjectToArray_0_, 0, aobject, 0, i);
             aobject[i] = p_addObjectToArray_1_;
             return aobject;
@@ -1922,7 +1922,7 @@
     {
         List list = new ArrayList<>(Arrays.asList(p_addObjectToArray_0_));
         list.add(p_addObjectToArray_2_, p_addObjectToArray_1_);
-        Object[] aobject = Array.newInstance(p_addObjectToArray_0_.getClass().getComponentType(), list.size());
+        Object[] aobject = (Object[]) Array.newInstance(p_addObjectToArray_0_.getClass().getComponentType(), list.size());
         return list.toArray(aobject);
     }
 
@@ -1940,7 +1940,7 @@
         {
             int i = p_addObjectsToArray_0_.length;
             int j = i + p_addObjectsToArray_1_.length;
-            Object[] aobject = Array.newInstance(p_addObjectsToArray_0_.getClass().getComponentType(), j);
+            Object[] aobject = (Object[]) Array.newInstance(p_addObjectsToArray_0_.getClass().getComponentType(), j);
             System.arraycopy(p_addObjectsToArray_0_, 0, aobject, 0, i);
             System.arraycopy(p_addObjectsToArray_1_, 0, aobject, i, p_addObjectsToArray_1_.length);
             return aobject;
