--- a/net/minecraft/client/renderer/debug/CollisionBoxDebugRenderer.java
+++ b/net/minecraft/client/renderer/debug/CollisionBoxDebugRenderer.java
@@ -32,9 +32,9 @@
             this.collisionData = activerenderinfo.func_216773_g().world.getCollisionShapes(activerenderinfo.func_216773_g(), activerenderinfo.func_216773_g().getBoundingBox().grow(6.0D), Collections.emptySet()).collect(Collectors.toList());
         }
 
-        double d1 = activerenderinfo.getProjectedView().x;
-        double d2 = activerenderinfo.getProjectedView().y;
-        double d3 = activerenderinfo.getProjectedView().z;
+        double d1 = activerenderinfo.getPosition().x;
+        double d2 = activerenderinfo.getPosition().y;
+        double d3 = activerenderinfo.getPosition().z;
         GlStateManager.enableBlend();
         GlStateManager.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
         GlStateManager.lineWidth(2.0F);
