--- a/net/minecraft/client/renderer/debug/EntityAIDebugRenderer.java
+++ b/net/minecraft/client/renderer/debug/EntityAIDebugRenderer.java
@@ -35,7 +35,7 @@
         GlStateManager.enableBlend();
         GlStateManager.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
         GlStateManager.disableTexture();
-        BlockPos blockpos = new BlockPos(activerenderinfo.getProjectedView().x, 0.0D, activerenderinfo.getProjectedView().z);
+        BlockPos blockpos = new BlockPos(activerenderinfo.getPosition().x, 0.0D, activerenderinfo.getPosition().z);
         this.field_217685_b.forEach((p_217683_1_, p_217683_2_) ->
         {
             for (int i = 0; i < p_217683_2_.size(); ++i)
