--- a/net/minecraft/client/renderer/debug/WaterDebugRenderer.java
+++ b/net/minecraft/client/renderer/debug/WaterDebugRenderer.java
@@ -21,9 +21,9 @@
     public void render(long p_217676_1_)
     {
         ActiveRenderInfo activerenderinfo = this.minecraft.gameRenderer.getActiveRenderInfo();
-        double d0 = activerenderinfo.getProjectedView().x;
-        double d1 = activerenderinfo.getProjectedView().y;
-        double d2 = activerenderinfo.getProjectedView().z;
+        double d0 = activerenderinfo.getPosition().x;
+        double d1 = activerenderinfo.getPosition().y;
+        double d2 = activerenderinfo.getPosition().z;
         BlockPos blockpos = this.minecraft.player.getPosition();
         IWorldReader iworldreader = this.minecraft.player.world;
         GlStateManager.enableBlend();
