--- a/net/minecraft/entity/Entity.java
+++ b/net/minecraft/entity/Entity.java
@@ -1720,7 +1720,7 @@
         }
     }
 
-    public RayTraceResult func_213324_a(double p_213324_1_, float p_213324_3_, boolean p_213324_4_)
+    public RayTraceResult rayTraceBlocksFromView(double p_213324_1_, float p_213324_3_, boolean p_213324_4_)
     {
         Vec3d vec3d = this.getEyePosition(p_213324_3_);
         Vec3d vec3d1 = this.getLook(p_213324_3_);
@@ -3265,7 +3265,7 @@
         return this.getEyeHeight(p_213307_1_, this.getSize(p_213307_1_));
     }
 
-    public final float getEyeHeight()
+    public float getEyeHeight()
     {
         return this.eyeHeight;
     }
