--- a/net/minecraft/client/gui/GuiVideoSettings$1.java
+++ b/net/minecraft/client/gui/GuiVideoSettings$1.java
@@ -5,7 +5,6 @@
     GuiVideoSettings$1(GuiVideoSettings p_i48008_1_, int p_i48008_2_, int p_i48008_3_, int p_i48008_4_, String p_i48008_5_)
     {
         super(p_i48008_2_, p_i48008_3_, p_i48008_4_, p_i48008_5_);
-        this.field_194926_o = p_i48008_1_;
     }
 
     /**
@@ -13,8 +12,5 @@
      */
     public void onClick(double mouseX, double mouseY)
     {
-        this.field_194926_o.mc.gameSettings.saveOptions();
-        this.field_194926_o.mc.mainWindow.update();
-        this.field_194926_o.mc.displayGuiScreen(GuiVideoSettings.access$000(this.field_194926_o));
     }
 }
