--- a/net/optifine/CustomColors.java
+++ b/net/optifine/CustomColors.java
@@ -337,7 +337,7 @@
 
         if (ainteger.length <= 0)
         {
-            return new ImmutablePair<LightMapPack[], Integer>((Object)null, 0);
+            return new ImmutablePair<LightMapPack[], Integer>((LightMapPack[])null, 0);
         }
         else
         {
@@ -476,7 +476,7 @@
         String s = "palette.block.";
         Map map = new HashMap();
 
-        for (String s1 : props.keySet())
+        for (String s1 : (Set<String>)(Set<?>)props.keySet())
         {
             String s2 = props.getProperty(s1);
 
@@ -630,15 +630,15 @@
             list.add((Object)null);
         }
 
-        List list = (List)list.get(id);
+        List sublist = (List)list.get(id);
 
-        if (list == null)
+        if (sublist == null)
         {
-            list = new ArrayList();
-            list.set(id, list);
+            sublist = new ArrayList();
+            list.set(id, sublist);
         }
 
-        list.add(cm);
+        sublist.add(cm);
     }
 
     private static CustomColormap[][] blockListToArray(List list)
@@ -647,11 +647,11 @@
 
         for (int i = 0; i < list.size(); ++i)
         {
-            List list = (List)list.get(i);
+            List sublist = (List)list.get(i);
 
-            if (list != null)
+            if (sublist != null)
             {
-                CustomColormap[] acustomcolormap1 = (CustomColormap[])list.toArray(new CustomColormap[list.size()]);
+                CustomColormap[] acustomcolormap1 = (CustomColormap[])sublist.toArray(new CustomColormap[sublist.size()]);
                 acustomcolormap[i] = acustomcolormap1;
             }
         }
@@ -1385,7 +1385,7 @@
         Set set = props.keySet();
         int i = 0;
 
-        for (String s : set)
+        for (String s : (Set<String>)(Set<?>)set)
         {
             String s1 = props.getProperty(s);
 
@@ -1529,7 +1529,7 @@
         float[][] afloat1 = new float[aenumdyecolor.length][];
         int k = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1602,7 +1602,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1658,7 +1658,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1697,7 +1697,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
