--- a/net/optifine/CustomColors.java
+++ b/net/optifine/CustomColors.java
@@ -467,7 +467,7 @@
         String s = "palette.block.";
         Map map = new HashMap();
 
-        for (String s1 : props.keySet())
+        for (String s1 : (Set<String>)(Set<?>)props.keySet())
         {
             String s2 = props.getProperty(s1);
 
@@ -477,7 +477,7 @@
             }
         }
 
-        String[] astring = map.keySet().toArray(new String[map.size()]);
+        String[] astring = (String[]) map.keySet().toArray(new String[map.size()]);
 
         for (int j = 0; j < astring.length; ++j)
         {
@@ -521,7 +521,7 @@
         }
         else
         {
-            CustomColormap[] acustomcolormap = list.toArray(new CustomColormap[list.size()]);
+            CustomColormap[] acustomcolormap = (CustomColormap[]) list.toArray(new CustomColormap[list.size()]);
             return acustomcolormap;
         }
     }
@@ -621,15 +621,15 @@
             list.add((Object)null);
         }
 
-        List list = (List)list.get(id);
+        List sublist = (List)list.get(id);
 
-        if (list == null)
+        if (sublist == null)
         {
-            list = new ArrayList();
-            list.set(id, list);
+            sublist = new ArrayList();
+            list.set(id, sublist);
         }
 
-        list.add(cm);
+        sublist.add(cm);
     }
 
     private static CustomColormap[][] blockListToArray(List list)
@@ -638,11 +638,11 @@
 
         for (int i = 0; i < list.size(); ++i)
         {
-            List list = (List)list.get(i);
+            List sublist = (List)list.get(i);
 
-            if (list != null)
+            if (sublist != null)
             {
-                CustomColormap[] acustomcolormap1 = list.toArray(new CustomColormap[list.size()]);
+                CustomColormap[] acustomcolormap1 = (CustomColormap[]) sublist.toArray(new CustomColormap[sublist.size()]);
                 acustomcolormap[i] = acustomcolormap1;
             }
         }
@@ -1376,7 +1376,7 @@
         Set set = props.keySet();
         int i = 0;
 
-        for (String s : set)
+        for (String s : (Set<String>)(Set<?>)set)
         {
             String s1 = props.getProperty(s);
 
@@ -1520,7 +1520,7 @@
         float[][] afloat1 = new float[adyecolor.length][];
         int k = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1593,7 +1593,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1649,7 +1649,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
@@ -1688,7 +1688,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = props.getProperty(s);
 
