--- a/net/optifine/CustomItems.java
+++ b/net/optifine/CustomItems.java
@@ -513,11 +513,11 @@
 
         for (int i = 0; i < list.size(); ++i)
         {
-            List list = (List)list.get(i);
+            List subList = (List)list.get(i);
 
-            if (list != null)
+            if (subList != null)
             {
-                CustomItemProperties[] acustomitemproperties1 = list.toArray(new CustomItemProperties[list.size()]);
+                CustomItemProperties[] acustomitemproperties1 = (CustomItemProperties[]) subList.toArray(new CustomItemProperties[subList.size()]);
                 Arrays.sort(acustomitemproperties1, new CustomItemsComparator());
                 acustomitemproperties[i] = acustomitemproperties1;
             }
@@ -589,15 +589,15 @@
             list.add((List<CustomItemProperties>)null);
         }
 
-        List<CustomItemProperties> list = list.get(id);
+        List<CustomItemProperties> subList = list.get(id);
 
-        if (list == null)
+        if (subList == null)
         {
-            list = new ArrayList<>();
-            list.set(id, list);
+        	subList = new ArrayList<>();
+            list.set(id, subList);
         }
 
-        list.add(cp);
+        subList.add(cp);
     }
 
     public static IBakedModel getCustomItemModel(ItemStack itemStack, IBakedModel model, ResourceLocation modelLocation, boolean fullModel)
