--- a/net/optifine/entity/model/CustomEntityModels.java
+++ b/net/optifine/entity/model/CustomEntityModels.java
@@ -19,6 +19,8 @@
 import net.minecraft.client.renderer.model.Model;
 import net.minecraft.client.renderer.tileentity.TileEntityRenderer;
 import net.minecraft.client.renderer.tileentity.TileEntityRendererDispatcher;
+import net.minecraft.entity.Entity;
+import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.ResourceLocation;
 import net.optifine.Config;
 import net.optifine.entity.model.anim.ModelResolver;
@@ -49,7 +51,7 @@
             map.clear();
             map1.clear();
             map.putAll(originalEntityRenderMap);
-            map1.putAll(originalTileEntityRenderMap);
+            map1.putAll((Map<? extends Class, ? extends TileEntityRenderer>) originalTileEntityRenderMap);
 
             if (Config.isCustomEntityModels())
             {
