--- a/net/minecraft/network/play/client/CPacketCustomPayload.java
+++ b/net/minecraft/network/play/client/CPacketCustomPayload.java
@@ -11,8 +11,8 @@
 {
     /** Channel for the client brand. */
     public static final ResourceLocation BRAND = new ResourceLocation("minecraft:brand");
-    private ResourceLocation channel;
-    private PacketBuffer data;
+    public ResourceLocation channel;
+    public PacketBuffer data;
 
     public CPacketCustomPayload()
     {
