--- a/net/minecraft/resources/AsyncReloader.java
+++ b/net/minecraft/resources/AsyncReloader.java
@@ -56,7 +56,7 @@
                             AsyncReloader.this.allAsyncCompleted.complete(Unit.INSTANCE);
                         }
                     });
-                    return AsyncReloader.this.allAsyncCompleted.thenCombine(completablefuture_f, (p_216874_1_, p_216874_2_) ->
+                    return AsyncReloader.this.allAsyncCompleted.thenCombine(completablefuture, (p_216874_1_, p_216874_2_) ->
                     {
                         return backgroundResult;
                     });
@@ -77,7 +77,6 @@
                 });
             });
             list.add(completablefuture1);
-            completablefuture = completablefuture1;
         }
 
         this.resultListFuture = Util.gather(list);
