--- a/net/optifine/util/FontUtils.java
+++ b/net/optifine/util/FontUtils.java
@@ -6,6 +6,7 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.Properties;
+import java.util.Set;
 import net.minecraft.src.Config;
 import net.minecraft.util.ResourceLocation;
 
@@ -55,7 +56,7 @@
     {
         Char2ObjectMap<Float> char2objectmap = new Char2ObjectOpenHashMap<Float>();
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String s1 = "width.";
 
