--- a/net/minecraft/command/impl/ForceLoadCommand.java
+++ b/net/minecraft/command/impl/ForceLoadCommand.java
@@ -10,6 +10,8 @@
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.brigadier.exceptions.Dynamic2CommandExceptionType.Function;
 import it.unimi.dsi.fastutil.longs.LongSet;
+
+import java.util.function.Consumer;
 import java.util.function.LongConsumer;
 import java.util.function.Predicate;
 import net.minecraft.command.CommandSource;
@@ -112,7 +114,7 @@
         DimensionType dimensiontype = p_502022_0_.getWorld().getDimension().getType();
         WorldServer worldserver = p_502022_0_.getServer().getWorld(dimensiontype);
         LongSet longset = worldserver.func_502116_ag();
-        longset.forEach((LongConsumer)(p_502024_1_) ->
+        longset.forEach((Consumer<? super Long>)(p_502024_1_) ->
         {
             worldserver.func_502118_b(ChunkPos.func_502104_a(p_502024_1_), ChunkPos.func_502105_b(p_502024_1_), false);
         });
