--- a/net/minecraft/client/particle/ParticleManager.java
+++ b/net/minecraft/client/particle/ParticleManager.java
@@ -397,9 +397,9 @@
         float f2 = -f1 * MathHelper.sin(p_215233_1_.getPitch() * ((float)Math.PI / 180F));
         float f3 = f * MathHelper.sin(p_215233_1_.getPitch() * ((float)Math.PI / 180F));
         float f4 = MathHelper.cos(p_215233_1_.getPitch() * ((float)Math.PI / 180F));
-        Particle.interpPosX = p_215233_1_.getProjectedView().x;
-        Particle.interpPosY = p_215233_1_.getProjectedView().y;
-        Particle.interpPosZ = p_215233_1_.getProjectedView().z;
+        Particle.interpPosX = p_215233_1_.getPosition().x;
+        Particle.interpPosY = p_215233_1_.getPosition().y;
+        Particle.interpPosZ = p_215233_1_.getPosition().z;
 
         for (IParticleRenderType iparticlerendertype : TYPES)
         {
