--- a/net/minecraft/client/particle/ParticleItemPickup.java
+++ b/net/minecraft/client/particle/ParticleItemPickup.java
@@ -7,6 +7,7 @@
 import net.minecraft.client.renderer.entity.RenderManager;
 import net.minecraft.entity.Entity;
 import net.minecraft.src.Config;
+import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 import net.optifine.shaders.Program;
 import net.optifine.shaders.Shaders;
@@ -50,9 +51,20 @@
         double d3 = this.target.lastTickPosX + (this.target.posX - this.target.lastTickPosX) * (double)partialTicks;
         double d4 = this.target.lastTickPosY + (this.target.posY - this.target.lastTickPosY) * (double)partialTicks + (double)this.yOffset;
         double d5 = this.target.lastTickPosZ + (this.target.posZ - this.target.lastTickPosZ) * (double)partialTicks;
+     
+        //Vivecraft
+        Minecraft mc = Minecraft.getMinecraft();
+        if (target== mc.player) {
+        	Vec3d pos = mc.entityRenderer.getControllerRenderPos(0);
+        	d3 = pos.x;
+        	d4 = pos.y;
+        	d5 = pos.z;
+        }
+        
         double d6 = d0 + (d3 - d0) * (double)f;
         double d7 = d1 + (d4 - d1) * (double)f;
         double d8 = d2 + (d5 - d2) * (double)f;
+        
         int i = this.getBrightnessForRender(partialTicks);
         int j = i % 65536;
         int k = i / 65536;
