--- a/net/minecraft/client/gui/inventory/GuiContainerCreative.java
+++ b/net/minecraft/client/gui/inventory/GuiContainerCreative.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.gui.inventory;
 
 import com.google.common.collect.Lists;
+import com.mtbs3d.minecrift.utils.ASMDelegator;
+
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -445,7 +447,11 @@
         {
             guicontainercreative$containercreative.itemList.addAll(this.mc.<ItemStack>getSearchTree(SearchTreeManager.ITEMS).search(this.searchField.getText().toLowerCase(Locale.ROOT)));
         }
-
+        
+        //VIVECRAFT idk what this is for anymore.
+        ASMDelegator.addCreativeSearch(this.searchField.getText(), guicontainercreative$containercreative.itemList);
+        //
+        
         this.currentScroll = 0.0F;
         guicontainercreative$containercreative.scrollTo(0.0F);
     }
@@ -565,6 +571,9 @@
         else if (tab != ItemGroup.SEARCH)
         {
             tab.fill(guicontainercreative$containercreative.itemList);
+            //VIVECRAFT - what is this even
+            ASMDelegator.addCreativeItems(tab, guicontainercreative$containercreative.itemList);
+            //
         }
 
         if (tab == ItemGroup.INVENTORY)
@@ -715,6 +724,8 @@
         this.drawDefaultBackground();
         super.render(mouseX, mouseY, partialTicks);
 
+        //VIVECRAFT ?? There used to be a check for mousedown here... where u go?
+        
         for (ItemGroup itemgroup : ItemGroup.GROUPS)
         {
             if (this.renderCreativeInventoryHoveringText(itemgroup, mouseX, mouseY))
