--- a/net/minecraft/client/gui/inventory/GuiEditSign.java
+++ b/net/minecraft/client/gui/inventory/GuiEditSign.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.gui.inventory;
 
+import com.mtbs3d.minecrift.gameplay.screenhandlers.KeyboardHandler;
+
 import net.minecraft.block.BlockStandingSign;
 import net.minecraft.block.BlockWallSign;
 import net.minecraft.block.state.IBlockState;
@@ -50,6 +52,7 @@
             }
         });
         this.tileSign.setEditable(false);
+        KeyboardHandler.setOverlayShowing(true);
     }
 
     /**
@@ -66,6 +69,7 @@
         }
 
         this.tileSign.setEditable(true);
+        KeyboardHandler.setOverlayShowing(false);
     }
 
     /**
