--- a/net/minecraft/client/renderer/entity/layers/HeldItemLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/HeldItemLayer.java
@@ -1,14 +1,22 @@
 package net.minecraft.client.renderer.entity.layers;
 
+import org.vivecraft.gameplay.trackers.BowTracker;
+import org.vivecraft.render.PlayerModelController;
+import org.vivecraft.render.PlayerModelController.RotInfo;
+
 import com.mojang.blaze3d.platform.GlStateManager;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.entity.IEntityRenderer;
 import net.minecraft.client.renderer.entity.model.EntityModel;
 import net.minecraft.client.renderer.entity.model.IHasArm;
+import net.minecraft.client.renderer.entity.model.VRPlayerModel;
 import net.minecraft.client.renderer.model.ItemCameraTransforms;
 import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemStack;
+import net.minecraft.util.Hand;
 import net.minecraft.util.HandSide;
+import net.minecraft.util.math.Vec3d;
 
 public class HeldItemLayer<T extends LivingEntity, M extends EntityModel<T> & IHasArm> extends LayerRenderer<T, M>
 {
@@ -42,23 +50,92 @@
 
     private void renderHeldItem(LivingEntity p_188358_1_, ItemStack p_188358_2_, ItemCameraTransforms.TransformType p_188358_3_, HandSide handSide)
     {
-        if (!p_188358_2_.isEmpty())
-        {
-            GlStateManager.pushMatrix();
-            this.translateToHand(handSide);
-
-            if (p_188358_1_.func_213287_bg())
-            {
-                GlStateManager.translatef(0.0F, 0.2F, 0.0F);
-            }
-
-            GlStateManager.rotatef(-90.0F, 1.0F, 0.0F, 0.0F);
-            GlStateManager.rotatef(180.0F, 0.0F, 1.0F, 0.0F);
-            boolean flag = handSide == HandSide.LEFT;
-            GlStateManager.translatef((float)(flag ? -1 : 1) / 16.0F, 0.125F, -0.625F);
-            Minecraft.getInstance().getFirstPersonRenderer().renderItemSide(p_188358_1_, p_188358_2_, p_188358_3_, flag);
-            GlStateManager.popMatrix();
-        }
+    	if (!p_188358_2_.isEmpty())
+    	{
+    		boolean vr =false;;
+    		RotInfo rotinfo = null;
+    		if(p_188358_1_ instanceof PlayerEntity){
+    			if(PlayerModelController.getInstance().isTracked(p_188358_1_.getUniqueID())){
+    				rotinfo = PlayerModelController.getInstance().getRotationsForPlayer(p_188358_1_.getUniqueID());
+    				if(rotinfo.seated ==false){
+    					vr = true;
+    				}
+    			}
+    		}
+
+    		if(handSide == HandSide.LEFT && vr && BowTracker.isHoldingBowEither(p_188358_1_)){
+    			return;
+    		}
+
+    		GlStateManager.pushMatrix();
+
+    		if (p_188358_1_.isSneaking())
+    		{
+    			GlStateManager.translatef(0.0F, 0.2F, 0.0F);
+    		}
+
+    		this.translateToHand(handSide);
+
+    		GlStateManager.rotatef(-90.0F, 1.0F, 0.0F, 0.0F);
+    		GlStateManager.rotatef(180.0F, 0.0F, 1.0F, 0.0F);
+    		boolean flag = handSide == HandSide.LEFT;
+
+    		//VIVECRAFT
+    		if(!vr) {
+    			GlStateManager.translatef((float)(flag ? -1 : 1) / 16.0F, 0.125F, -0.625F);   
+    		}
+    		else{
+    			GlStateManager.translatef((float)(flag ? -1 : 1) / 16.0F, 0.125F,  -.2125f);
+    			if(BowTracker.isBow(p_188358_2_) && !rotinfo.seated){
+
+    				Vec3d diff = rotinfo.leftArmPos.subtract(rotinfo.rightArmPos);
+    				double ltor = Math.atan2(-diff.x, diff.z) +  Math.PI/2;   
+    				Vec3d rtol = rotinfo.rightArmPos.subtract(rotinfo.leftArmPos).rotateYaw((float) ltor);
+    				Vec3d larm = rotinfo.leftArmPos;
+    				VRPlayerModel m = (VRPlayerModel) this.getEntityModel();
+    				larm = larm.rotateYaw((float)(ltor)).add(rotinfo.leftArmRot.scale(-0.2));  
+    				float scale = (float) 0.0625;
+
+    				GlStateManager.popMatrix();
+    				GlStateManager.pushMatrix();
+
+    				GlStateManager.translatef(m.bipedLeftArm.rotationPointX*scale,m.bipedLeftArm.rotationPointY*scale,m.bipedLeftArm.rotationPointZ*scale);
+
+    				if(p_188358_1_.isHandActive()){
+
+    					float yaw1 = (float) Math.atan2(-rtol.x, -rtol.z); 
+    					float pitch1 = (float) Math.asin(rtol.y/rtol.length()); 
+
+    					GlStateManager.rotatef((float) Math.toDegrees(-Math.PI + yaw1), 0,1,0);
+    					GlStateManager.rotatef((float) Math.toDegrees(pitch1), 1,0,0);
+
+    				} else{
+    					GlStateManager.rotatef((float) Math.toDegrees(m.bipedLeftArm.rotateAngleY), 0,1,0);
+    					GlStateManager.rotatef( 90 + (float) Math.toDegrees(m.bipedLeftArm.rotateAngleX), 1,0,0);
+    				}
+
+    				GlStateManager.translatef(0, 0.155F, -0.215F);
+    				Minecraft.getInstance().getFirstPersonRenderer().renderItemSide(p_188358_1_, p_188358_2_, p_188358_3_, flag);
+    				//
+    				GlStateManager.popMatrix();
+ 
+    				GlStateManager.pushMatrix();
+    				this.translateToHand(handSide);
+    				GlStateManager.popMatrix();
+    				return;
+    			} else {
+    				GlStateManager.rotatef(-80.0F, 1.0F, 0.0F, 0.0F);        		
+    			}
+
+
+    			GlStateManager.translatef(0, 0.155F, -0.215F);
+
+    		}
+
+    		//
+    		Minecraft.getInstance().getFirstPersonRenderer().renderItemSide(p_188358_1_, p_188358_2_, p_188358_3_, flag);
+    		GlStateManager.popMatrix();
+    	}
     }
 
     protected void translateToHand(HandSide p_191361_1_)
