--- a/net/optifine/gui/GuiScreenCapeOF.java
+++ b/net/optifine/gui/GuiScreenCapeOF.java
@@ -17,10 +17,10 @@
     private String message;
     private long messageHideTimeMs;
 
-    public GuiScreenCapeOF(Screen parentScreenIn)
+    public GuiScreenCapeOF(Screen parentGuiScreenIn)
     {
         super(new StringTextComponent(I18n.format("of.options.capeOF.title")));
-        this.parentScreen = parentScreenIn;
+        this.parentScreen = parentGuiScreenIn;
     }
 
     protected void init()
