--- a/net/optifine/gui/GuiScreenCapeOF.java
+++ b/net/optifine/gui/GuiScreenCapeOF.java
@@ -13,14 +13,14 @@
 
 public class GuiScreenCapeOF extends GuiScreenOF
 {
-    private final Screen parentScreen;
+    private final Screen parentGuiScreen;
     private String message;
     private long messageHideTimeMs;
 
-    public GuiScreenCapeOF(Screen parentScreenIn)
+    public GuiScreenCapeOF(Screen parentGuiScreenIn)
     {
         super(new StringTextComponent(I18n.format("of.options.capeOF.title")));
-        this.parentScreen = parentScreenIn;
+        this.parentGuiScreen = parentGuiScreenIn;
     }
 
     protected void init()
@@ -43,7 +43,7 @@
             {
                 if (guibuttonof.id == 200)
                 {
-                    this.minecraft.displayGuiScreen(this.parentScreen);
+                    this.minecraft.displayGuiScreen(this.parentGuiScreen);
                 }
 
                 if (guibuttonof.id == 210)
