--- a/net/minecraft/resources/VanillaPack.java
+++ b/net/minecraft/resources/VanillaPack.java
@@ -31,6 +31,8 @@
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.Util;
 import net.optifine.reflect.ReflectorForge;
+
+import org.apache.commons.io.IOUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -303,11 +305,16 @@
         try
         {
             URL url = VanillaPack.class.getResource(s);
-            return func_223459_a(s, url);
+            boolean flag = func_223459_a(s, url);
+            if (!flag) throw new IOException();
+            return flag;
         }
         catch (IOException var51)
         {
-            return false;
+            InputStream inputstream = this.getInputStreamVanilla(type, location);
+            boolean flag = inputstream != null;
+            IOUtils.closeQuietly(inputstream);
+            return flag;
         }
     }
 
