--- a/net/minecraft/client/renderer/ActiveRenderInfo.java
+++ b/net/minecraft/client/renderer/ActiveRenderInfo.java
@@ -18,16 +18,16 @@
 
 public class ActiveRenderInfo
 {
-    private boolean valid;
-    private IBlockReader world;
-    private Entity renderViewEntity;
+    protected boolean valid;
+    protected IBlockReader world;
+    protected Entity renderViewEntity;
     private Vec3d pos = Vec3d.ZERO;
     private final BlockPos.MutableBlockPos blockPos = new BlockPos.MutableBlockPos();
-    private Vec3d look;
-    private Vec3d up;
-    private Vec3d field_216796_h;
-    private float pitch;
-    private float yaw;
+    protected Vec3d look;
+    protected Vec3d up;
+    protected Vec3d field_216796_h;
+    protected float pitch;
+    protected float yaw;
     private boolean thirdPerson;
     private boolean thirdPersonReverse;
     private float height;
