--- a/net/minecraft/client/renderer/ActiveRenderInfo.java
+++ b/net/minecraft/client/renderer/ActiveRenderInfo.java
@@ -17,16 +17,16 @@
 
 public class ActiveRenderInfo
 {
-    private boolean valid;
-    private IBlockReader world;
-    private Entity field_216791_c;
+    protected boolean valid;
+    protected IBlockReader world;
+    protected Entity field_216791_c;
     private Vec3d pos = Vec3d.ZERO;
     private final BlockPos.MutableBlockPos blockPos = new BlockPos.MutableBlockPos();
-    private Vec3d look;
-    private Vec3d up;
-    private Vec3d field_216796_h;
-    private float pitch;
-    private float yaw;
+    protected Vec3d look;
+    protected Vec3d up;
+    protected Vec3d field_216796_h;
+    protected float pitch;
+    protected float yaw;
     private boolean thirdPerson;
     private boolean thirdPersonReverse;
     private float height;
@@ -161,7 +161,7 @@
         return this.pos;
     }
 
-    public BlockPos func_216780_d()
+    public BlockPos getBlockPos()
     {
         return this.blockPos;
     }
@@ -181,7 +181,7 @@
         return this.field_216791_c;
     }
 
-    public boolean func_216786_h()
+    public boolean isValid()
     {
         return this.valid;
     }
