--- a/net/minecraft/client/renderer/FogRenderer.java
+++ b/net/minecraft/client/renderer/FogRenderer.java
@@ -36,12 +36,12 @@
     private int lastWaterFogColor = -1;
     private int waterFogColor = -1;
     private long waterFogUpdateTime = -1L;
-    private final GameRenderer entityRenderer;
+    private final GameRenderer gameRenderer;
     private final Minecraft mc;
 
     public FogRenderer(GameRenderer p_i91_1_)
     {
-        this.entityRenderer = p_i91_1_;
+        this.gameRenderer = p_i91_1_;
         this.mc = p_i91_1_.getMinecraft();
         this.blackBuffer.put(0.0F).put(0.0F).put(0.0F).put(1.0F).flip();
     }
@@ -97,9 +97,9 @@
             this.blue = (float)((double)this.blue * d0);
         }
 
-        if (this.entityRenderer.getBossColorModifier(p_217619_2_) > 0.0F)
+        if (this.gameRenderer.getBossColorModifier(p_217619_2_) > 0.0F)
         {
-            float f = this.entityRenderer.getBossColorModifier(p_217619_2_);
+            float f = this.gameRenderer.getBossColorModifier(p_217619_2_);
             this.red = this.red * (1.0F - f) + this.red * 0.7F * f;
             this.green = this.green * (1.0F - f) + this.green * 0.6F * f;
             this.blue = this.blue * (1.0F - f) + this.blue * 0.6F * f;
@@ -138,7 +138,7 @@
         }
         else if (p_217619_1_.getRenderViewEntity() instanceof LivingEntity && ((LivingEntity)p_217619_1_.getRenderViewEntity()).isPotionActive(Effects.NIGHT_VISION))
         {
-            float f2 = this.entityRenderer.getNightVisionBrightness((LivingEntity)p_217619_1_.getRenderViewEntity(), p_217619_2_);
+            float f2 = this.gameRenderer.getNightVisionBrightness((LivingEntity)p_217619_1_.getRenderViewEntity(), p_217619_2_);
             float f4 = 1.0F / this.red;
 
             if (f4 > 1.0F / this.green)
@@ -186,7 +186,7 @@
 
         if (Reflector.EntityViewRenderEvent_FogColors_Constructor.exists())
         {
-            Object object = Reflector.newInstance(Reflector.EntityViewRenderEvent_FogColors_Constructor, this, this.entityRenderer, p_217619_1_, p_217619_2_, this.red, this.green, this.blue);
+            Object object = Reflector.newInstance(Reflector.EntityViewRenderEvent_FogColors_Constructor, this, this.gameRenderer, p_217619_1_, p_217619_2_, this.red, this.green, this.blue);
             Reflector.postForgeBusEvent(object);
             this.red = Reflector.callFloat(object, Reflector.EntityViewRenderEvent_FogColors_getRed);
             this.green = Reflector.callFloat(object, Reflector.EntityViewRenderEvent_FogColors_getGreen);
@@ -302,7 +302,7 @@
 
     public void setupFog(ActiveRenderInfo p_setupFog_1_, int p_setupFog_2_, float p_setupFog_3_)
     {
-        this.entityRenderer.fogStandard = false;
+        this.gameRenderer.fogStandard = false;
         this.applyFog(false);
         GlStateManager.normal3f(0.0F, -1.0F, 0.0F);
         GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
@@ -312,7 +312,7 @@
 
         if (Reflector.ForgeHooksClient_getFogDensity.exists())
         {
-            f = Reflector.callFloat(Reflector.ForgeHooksClient_getFogDensity, this, this.entityRenderer, p_setupFog_1_, p_setupFog_3_, 0.1F);
+            f = Reflector.callFloat(Reflector.ForgeHooksClient_getFogDensity, this, this.gameRenderer, p_setupFog_1_, p_setupFog_3_, 0.1F);
         }
 
         if (f >= 0.0F)
@@ -326,7 +326,7 @@
 
             if (i < 20)
             {
-                f3 = MathHelper.lerp(1.0F - (float)i / 20.0F, 5.0F, this.entityRenderer.getFarPlaneDistance());
+                f3 = MathHelper.lerp(1.0F - (float)i / 20.0F, 5.0F, this.gameRenderer.getFarPlaneDistance());
             }
 
             GlStateManager.fogMode(GlStateManager.FogMode.LINEAR);
@@ -380,8 +380,8 @@
         }
         else
         {
-            float f2 = this.entityRenderer.getFarPlaneDistance();
-            this.entityRenderer.fogStandard = true;
+            float f2 = this.gameRenderer.getFarPlaneDistance();
+            this.gameRenderer.fogStandard = true;
             GlStateManager.fogMode(GlStateManager.FogMode.LINEAR);
 
             if (p_setupFog_2_ == -1)
@@ -405,7 +405,7 @@
 
             if (Reflector.ForgeHooksClient_onFogRender.exists())
             {
-                Reflector.callVoid(Reflector.ForgeHooksClient_onFogRender, this, this.entityRenderer, p_setupFog_1_, p_setupFog_3_, p_setupFog_2_, f2);
+                Reflector.callVoid(Reflector.ForgeHooksClient_onFogRender, this, this.gameRenderer, p_setupFog_1_, p_setupFog_3_, p_setupFog_2_, f2);
             }
         }
 
