--- a/net/minecraft/client/renderer/FogRenderer.java
+++ b/net/minecraft/client/renderer/FogRenderer.java
@@ -37,13 +37,13 @@
     private int lastWaterFogColor = -1;
     private int waterFogColor = -1;
     private long waterFogUpdateTime = -1L;
-    private final GameRenderer entityRenderer;
+    private final GameRenderer gameRenderer;
     private final Minecraft mc;
 
     public FogRenderer(GameRenderer p_i115_1_)
     {
-        this.entityRenderer = p_i115_1_;
-        this.mc = p_i115_1_.getMinecraft();
+        this.gameRenderer = p_i115_1_;
+        this.mc = p_i115_1_.getInstance();
         this.blackBuffer.put(0.0F).put(0.0F).put(0.0F).put(1.0F).flip();
     }
 
@@ -69,7 +69,7 @@
             this.waterFogUpdateTime = -1L;
         }
 
-        double d0 = p_217619_1_.getProjectedView().y * world.dimension.getVoidFogYFactor();
+        double d0 = p_217619_1_.getPosition().y * world.dimension.getVoidFogYFactor();
 
         if (p_217619_1_.func_216773_g() instanceof LivingEntity && ((LivingEntity)p_217619_1_.func_216773_g()).isPotionActive(Effects.BLINDNESS))
         {
@@ -98,9 +98,9 @@
             this.blue = (float)((double)this.blue * d0);
         }
 
-        if (this.entityRenderer.getBossColorModifier(p_217619_2_) > 0.0F)
+        if (this.gameRenderer.getBossColorModifier(p_217619_2_) > 0.0F)
         {
-            float f = this.entityRenderer.getBossColorModifier(p_217619_2_);
+            float f = this.gameRenderer.getBossColorModifier(p_217619_2_);
             this.red = this.red * (1.0F - f) + this.red * 0.7F * f;
             this.green = this.green * (1.0F - f) + this.green * 0.6F * f;
             this.blue = this.blue * (1.0F - f) + this.blue * 0.6F * f;
@@ -139,7 +139,7 @@
         }
         else if (p_217619_1_.func_216773_g() instanceof LivingEntity && ((LivingEntity)p_217619_1_.func_216773_g()).isPotionActive(Effects.NIGHT_VISION))
         {
-            float f2 = this.entityRenderer.getNightVisionBrightness((LivingEntity)p_217619_1_.func_216773_g(), p_217619_2_);
+            float f2 = this.gameRenderer.getNightVisionBrightness((LivingEntity)p_217619_1_.func_216773_g(), p_217619_2_);
             float f4 = 1.0F / this.red;
 
             if (f4 > 1.0F / this.green)
@@ -189,7 +189,7 @@
         {
             Entity entity = p_217619_1_.func_216773_g();
             BlockState blockstate = p_217619_1_.getBlockState();
-            Object object = Reflector.newInstance(Reflector.EntityViewRenderEvent_FogColors_Constructor, this, this.entityRenderer, entity, blockstate, ifluidstate, p_217619_2_, this.red, this.green, this.blue);
+            Object object = Reflector.newInstance(Reflector.EntityViewRenderEvent_FogColors_Constructor, this, this.gameRenderer, entity, blockstate, ifluidstate, p_217619_2_, this.red, this.green, this.blue);
             Reflector.postForgeBusEvent(object);
             this.red = Reflector.callFloat(object, Reflector.EntityViewRenderEvent_FogColors_getRed);
             this.green = Reflector.callFloat(object, Reflector.EntityViewRenderEvent_FogColors_getGreen);
@@ -203,7 +203,7 @@
     {
         float f = 0.25F + 0.75F * (float)this.mc.gameSettings.renderDistanceChunks / 32.0F;
         f = 1.0F - (float)Math.pow((double)f, 0.25D);
-        Vec3d vec3d = p_217620_2_.func_217382_a(p_217620_1_.func_216780_d(), p_217620_3_);
+        Vec3d vec3d = p_217620_2_.func_217382_a(p_217620_1_.getBlockPos(), p_217620_3_);
         vec3d = CustomColors.getWorldSkyColor(vec3d, p_217620_2_, this.mc.getRenderViewEntity(), p_217620_3_);
         float f1 = (float)vec3d.x;
         float f2 = (float)vec3d.y;
@@ -267,7 +267,7 @@
     private void func_217621_a(ActiveRenderInfo p_217621_1_, IWorldReader p_217621_2_)
     {
         long i = Util.milliTime();
-        int j = p_217621_2_.getBiome(new BlockPos(p_217621_1_.getProjectedView())).getWaterFogColor();
+        int j = p_217621_2_.getBiome(new BlockPos(p_217621_1_.getPosition())).getWaterFogColor();
 
         if (this.waterFogUpdateTime < 0L)
         {
@@ -300,7 +300,7 @@
 
     public void func_217618_a(ActiveRenderInfo p_217618_1_, int p_217618_2_)
     {
-        this.entityRenderer.fogStandard = false;
+        this.gameRenderer.fogStandard = false;
         this.applyFog(false);
         GlStateManager.normal3f(0.0F, -1.0F, 0.0F);
         GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
@@ -311,7 +311,7 @@
 
         if (Reflector.ForgeHooksClient_getFogDensity.exists())
         {
-            f = Reflector.callFloat(Reflector.ForgeHooksClient_getFogDensity, this, this.entityRenderer, entity, blockstate, ifluidstate, 0, 0.1F);
+            f = Reflector.callFloat(Reflector.ForgeHooksClient_getFogDensity, this, this.gameRenderer, entity, blockstate, ifluidstate, 0, 0.1F);
         }
 
         if (f >= 0.0F)
@@ -325,7 +325,7 @@
 
             if (i < 20)
             {
-                f3 = MathHelper.lerp(1.0F - (float)i / 20.0F, 5.0F, this.entityRenderer.getFarPlaneDistance());
+                f3 = MathHelper.lerp(1.0F - (float)i / 20.0F, 5.0F, this.gameRenderer.getFarPlaneDistance());
             }
 
             GlStateManager.fogMode(GlStateManager.FogMode.LINEAR);
@@ -379,8 +379,8 @@
         }
         else
         {
-            float f2 = this.entityRenderer.getFarPlaneDistance();
-            this.entityRenderer.fogStandard = true;
+            float f2 = this.gameRenderer.getFarPlaneDistance();
+            this.gameRenderer.fogStandard = true;
             GlStateManager.fogMode(GlStateManager.FogMode.LINEAR);
 
             if (p_217618_2_ == -1)
@@ -396,7 +396,7 @@
 
             GLX.setupNvFogDistance();
 
-            if (this.mc.world.dimension.doesXZShowFog(MathHelper.floor(p_217618_1_.getProjectedView().x), MathHelper.floor(p_217618_1_.getProjectedView().z)) || this.mc.ingameGUI.getBossOverlay().shouldCreateFog())
+            if (this.mc.world.dimension.doesXZShowFog(MathHelper.floor(p_217618_1_.getPosition().x), MathHelper.floor(p_217618_1_.getPosition().z)) || this.mc.ingameGUI.getBossOverlay().shouldCreateFog())
             {
                 GlStateManager.fogStart(f2 * 0.05F);
                 GlStateManager.fogEnd(f2);
@@ -404,7 +404,7 @@
 
             if (Reflector.ForgeHooksClient_onFogRender.exists())
             {
-                Reflector.callVoid(Reflector.ForgeHooksClient_onFogRender, this, this.entityRenderer, p_217618_1_.func_216773_g(), blockstate, ifluidstate, 0, p_217618_2_, f2);
+                Reflector.callVoid(Reflector.ForgeHooksClient_onFogRender, this, this.gameRenderer, p_217618_1_.func_216773_g(), blockstate, ifluidstate, 0, p_217618_2_, f2);
             }
         }
 
