--- a/net/minecraft/client/renderer/WorldRenderer.java
+++ b/net/minecraft/client/renderer/WorldRenderer.java
@@ -49,6 +49,7 @@
 import net.minecraft.client.renderer.entity.EntityRendererManager;
 import net.minecraft.client.renderer.entity.ItemFrameRenderer;
 import net.minecraft.client.renderer.model.ModelBakery;
+import net.minecraft.client.renderer.model.ModelResourceLocation;
 import net.minecraft.client.renderer.texture.AtlasTexture;
 import net.minecraft.client.renderer.texture.TextureAtlasSprite;
 import net.minecraft.client.renderer.texture.TextureManager;
@@ -126,6 +127,9 @@
 import net.optifine.util.RenderChunkUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.provider.MCOpenVR;
+import org.vivecraft.render.RenderPass;
+import org.vivecraft.settings.VRSettings;
 
 public class WorldRenderer implements AutoCloseable, IResourceManagerReloadListener
 {
@@ -143,7 +147,7 @@
     private Set<ChunkRender> chunksToUpdate = new ObjectLinkedOpenHashSet<>();
     private List<WorldRenderer.LocalRenderInformationContainer> renderInfos = Lists.newArrayListWithCapacity(69696);
     private final Set<TileEntity> setTileEntities = Sets.newHashSet();
-    private ViewFrustum viewFrustum;
+    public ViewFrustum viewFrustum;
     private int starGLCallList = -1;
     private int glSkyList = -1;
     private int glSkyList2 = -1;
@@ -297,9 +301,12 @@
 
             try
             {
-                this.entityOutlineShader = new ShaderGroup(this.mc.getTextureManager(), this.mc.getResourceManager(), this.mc.getFramebuffer(), resourcelocation);
-                this.entityOutlineShader.createBindFramebuffers(this.mc.mainWindow.getFramebufferWidth(), this.mc.mainWindow.getFramebufferHeight());
-                this.entityOutlineFramebuffer = this.entityOutlineShader.getFramebufferRaw("final");
+            	if(mc.getFramebuffer() != null) {
+	            	this.entityOutlineShader = new ShaderGroup(this.mc.getTextureManager(), this.mc.getResourceManager(), this.mc.getFramebuffer(), resourcelocation);
+	                // VIVECRAFT: Use eye framebuffer instead of window
+	                this.entityOutlineShader.createBindFramebuffers(this.mc.stereoProvider.framebufferVrRender.framebufferWidth, this.mc.stereoProvider.framebufferVrRender.framebufferHeight);
+	                this.entityOutlineFramebuffer = this.entityOutlineShader.getFramebufferRaw("final");
+            	}
             }
             catch (IOException ioexception)
             {
@@ -327,7 +334,7 @@
         {
             GlStateManager.enableBlend();
             GlStateManager.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ZERO, GlStateManager.DestFactor.ONE);
-            this.entityOutlineFramebuffer.framebufferRenderExt(this.mc.mainWindow.getFramebufferWidth(), this.mc.mainWindow.getFramebufferHeight(), false);
+            this.entityOutlineFramebuffer.framebufferRenderExt(this.mc.getFramebuffer().framebufferWidth, this.mc.getFramebuffer().framebufferHeight, false);
             GlStateManager.disableBlend();
         }
     }
@@ -720,7 +727,18 @@
 
             for (Entity entity : this.world.getAllEntities())
             {
-                if ((this.renderManager.shouldRender(entity, p_215326_2_, d0, d1, d2) || entity.isRidingOrBeingRiddenBy(this.mc.player)) && (entity != p_215326_1_.getRenderViewEntity() || Shaders.isShadowPass || p_215326_1_.isThirdPerson() || p_215326_1_.getRenderViewEntity() instanceof LivingEntity && ((LivingEntity)p_215326_1_.getRenderViewEntity()).isSleeping()))
+            	
+            	//Vivecraft
+            	boolean selfRender = mc.currentPass == RenderPass.THIRD && mc.vrSettings.displayMirrorMode == VRSettings.MIRROR_THIRD_PERSON;
+            	selfRender |= mc.vrSettings.shouldRenderSelf || mc.vrSettings.tmpRenderSelf;
+            	//		
+            	if ((this.renderManager.shouldRender(entity, p_215326_2_, d0, d1, d2) || 
+            							entity.isRidingOrBeingRiddenBy(this.mc.player)) &&
+            					(selfRender || entity != p_215326_1_.getRenderViewEntity() ||
+            						Shaders.isShadowPass ||
+            							p_215326_1_.isThirdPerson() || 
+            								p_215326_1_.getRenderViewEntity() instanceof LivingEntity && 
+            									((LivingEntity)p_215326_1_.getRenderViewEntity()).isSleeping()))
                 {
                     ++this.countEntitiesRendered;
                     this.renderedEntity = entity;
@@ -729,10 +747,21 @@
                     {
                         Shaders.nextEntity(entity);
                     }
-
+                    //Vivecraft
+                    if(entity == this.mc.getRenderViewEntity()) { //fix shaders shadow location.
+                    	mc.gameRenderer.restoreRVEPos((LivingEntity) entity);
+                    }
+                    //
                     this.renderManager.renderEntityStatic(entity, p_215326_3_, false);
                     this.renderedEntity = null;
 
+                    //Vivecraft
+                    if(entity == this.mc.getRenderViewEntity()) {
+                    	mc.gameRenderer.cacheRVEPos((LivingEntity) entity);
+                    	mc.gameRenderer.setupRVE();
+                    }
+                    //
+
                     if (entity.isGlowing() || entity instanceof PlayerEntity && this.mc.player.isSpectator() && this.mc.gameSettings.keyBindSpectatorOutlines.isKeyDown())
                     {
                         list.add(entity);
@@ -771,7 +800,8 @@
                 {
                     GlStateManager.depthFunc(519);
                     GlStateManager.disableFog();
-                    this.entityOutlineFramebuffer.bindFramebuffer(false);
+                    // VIVECRAFT: Set viewport to the shader framebuffer
+                    this.entityOutlineFramebuffer.bindFramebuffer(true);
                     RenderHelper.disableStandardItemLighting();
                     this.renderManager.setRenderOutlines(true);
 
@@ -803,7 +833,8 @@
                     GlStateManager.enableAlphaTest();
                 }
 
-                this.mc.getFramebuffer().bindFramebuffer(false);
+                // VIVECRAFT: Reset viewport to normal
+                this.mc.getFramebuffer().bindFramebuffer(true);
             }
 
             if (!this.isRenderEntityOutlines() && (!list.isEmpty() || this.entityOutlinesRendered))
@@ -1074,7 +1105,9 @@
         new BlockPos(MathHelper.floor(p_215320_1_.getProjectedView().x / 16.0D) * 16, MathHelper.floor(p_215320_1_.getProjectedView().y / 16.0D) * 16, MathHelper.floor(p_215320_1_.getProjectedView().z / 16.0D) * 16);
         float f = p_215320_1_.getPitch();
         float f1 = p_215320_1_.getYaw();
-        this.displayListEntitiesDirty = this.displayListEntitiesDirty || !this.chunksToUpdate.isEmpty() || p_215320_1_.getProjectedView().x != this.lastViewEntityX || p_215320_1_.getProjectedView().y != this.lastViewEntityY || p_215320_1_.getProjectedView().z != this.lastViewEntityZ || (double)f != this.lastViewEntityPitch || (double)f1 != this.lastViewEntityYaw;
+         //Vivecraft 
+        this.displayListEntitiesDirty = true; //this.displayListEntitiesDirty || !this.chunksToUpdate.isEmpty() || entityIn.posX != this.lastViewEntityX || entityIn.posY != this.lastViewEntityY || entityIn.posZ != this.lastViewEntityZ || (double)f != this.lastViewEntityPitch || (double)f1 != this.lastViewEntityYaw;
+        //
         this.lastViewEntityX = p_215320_1_.getProjectedView().x;
         this.lastViewEntityY = p_215320_1_.getProjectedView().y;
         this.lastViewEntityZ = p_215320_1_.getProjectedView().z;
@@ -1449,7 +1482,7 @@
         }
     }
 
-    private void renderBlockLayer(BlockRenderLayer blockLayerIn)
+    void renderBlockLayer(BlockRenderLayer blockLayerIn)
     {
         this.mc.gameRenderer.enableLightmap();
 
@@ -1626,7 +1659,7 @@
 
             if (flag)
             {
-                Shaders.disableTexture2D();
+                Shaders.disableTexture();
             }
 
             Vec3d vec3d = this.world.getSkyColor(this.mc.gameRenderer.getActiveRenderInfo().getBlockPos(), partialTicks);
@@ -1694,7 +1727,7 @@
 
                 if (flag)
                 {
-                    Shaders.disableTexture2D();
+                    Shaders.disableTexture();
                 }
 
                 GlStateManager.shadeModel(7425);
@@ -1726,7 +1759,7 @@
 
             if (flag)
             {
-                Shaders.enableTexture2D();
+                Shaders.enableTexture();
             }
 
             GlStateManager.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
@@ -1785,7 +1818,7 @@
 
             if (flag)
             {
-                Shaders.disableTexture2D();
+                Shaders.disableTexture();
             }
 
             float f12 = this.world.getStarBrightness(partialTicks) * f10;
@@ -1824,7 +1857,7 @@
 
             if (flag)
             {
-                Shaders.disableTexture2D();
+                Shaders.disableTexture();
             }
 
             GlStateManager.color3f(0.0F, 0.0F, 0.0F);
@@ -1894,7 +1927,7 @@
 
             if (flag)
             {
-                Shaders.enableTexture2D();
+                Shaders.enableTexture();
             }
 
             GlStateManager.depthMask(true);
@@ -2484,6 +2517,8 @@
         }
     }
 
+    public float selR, selG, selB;
+    
     public void drawSelectionBox(ActiveRenderInfo p_215325_1_, RayTraceResult p_215325_2_, int p_215325_3_)
     {
         if (p_215325_3_ == 0 && p_215325_2_.getType() == RayTraceResult.Type.BLOCK)
@@ -2500,7 +2535,7 @@
 
                 if (Config.isShaders())
                 {
-                    Shaders.disableTexture2D();
+                    Shaders.disableTexture();
                 }
 
                 GlStateManager.depthMask(false);
@@ -2510,7 +2545,7 @@
                 double d0 = p_215325_1_.getProjectedView().x;
                 double d1 = p_215325_1_.getProjectedView().y;
                 double d2 = p_215325_1_.getProjectedView().z;
-                drawShape(blockstate.getShape(this.world, blockpos, ISelectionContext.forEntity(p_215325_1_.getRenderViewEntity())), (double)blockpos.getX() - d0, (double)blockpos.getY() - d1, (double)blockpos.getZ() - d2, 0.0F, 0.0F, 0.0F, 0.4F);
+                drawShape(blockstate.getShape(this.world, blockpos, ISelectionContext.forEntity(p_215325_1_.getRenderViewEntity())), (double)blockpos.getX() - d0, (double)blockpos.getY() - d1, (double)blockpos.getZ() - d2, selR, selG, selB, 0.4F);
                 GlStateManager.popMatrix();
                 GlStateManager.matrixMode(5888);
                 GlStateManager.depthMask(true);
@@ -2518,7 +2553,7 @@
 
                 if (Config.isShaders())
                 {
-                    Shaders.enableTexture2D();
+                    Shaders.enableTexture();
                 }
 
                 GlStateManager.disableBlend();
@@ -2979,6 +3014,8 @@
     {
         Random random = this.world.rand;
 
+        boolean playernear =  mc.player !=null && mc.player.isAlive() && (mc.player.getPosition().distanceSq(blockPosIn) < 25); ///hmm sure why not.
+        
         switch (type)
         {
             case 1000:
@@ -3034,18 +3071,26 @@
                 break;
 
             case 1011:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_IRON_DOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1012:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_WOODEN_DOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1013:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_WOODEN_TRAPDOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1014:
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_FENCE_GATE_CLOSE, SoundCategory.BLOCKS, 1.0F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -3066,14 +3111,26 @@
                 break;
 
             case 1019:
+                if(playernear){ //this is just mean.
+                	MCOpenVR.triggerHapticPulse(0,750); 
+                	MCOpenVR.triggerHapticPulse(1,750); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.ENTITY_ZOMBIE_ATTACK_WOODEN_DOOR, SoundCategory.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
             case 1020:
+                if(playernear){ //this is just mean.
+                	MCOpenVR.triggerHapticPulse(0,750); 
+                	MCOpenVR.triggerHapticPulse(1,750); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.ENTITY_ZOMBIE_ATTACK_IRON_DOOR, SoundCategory.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
             case 1021:
+                if(playernear){ //this is just mean.
+                	MCOpenVR.triggerHapticPulse(0,750); 
+                	MCOpenVR.triggerHapticPulse(1,750); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.ENTITY_ZOMBIE_BREAK_WOODEN_DOOR, SoundCategory.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
@@ -3102,10 +3159,17 @@
                 break;
 
             case 1030:
+                if(playernear){ 
+                	MCOpenVR.triggerHapticPulse(0,500); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_ANVIL_USE, SoundCategory.BLOCKS, 1.0F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1031:
+                if(playernear){ 
+                	MCOpenVR.triggerHapticPulse(0,1250); 
+                	MCOpenVR.triggerHapticPulse(1,1250); 
+                }
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_ANVIL_LAND, SoundCategory.BLOCKS, 0.3F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -3126,6 +3190,9 @@
                 break;
 
             case 1036:
+
+                if(playernear)
+                	MCOpenVR.triggerHapticPulse(0,250); //VIVECRAFT go thump.
                 this.world.playSound(blockPosIn, SoundEvents.BLOCK_IRON_TRAPDOOR_CLOSE, SoundCategory.BLOCKS, 1.0F, this.world.rand.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -3482,4 +3549,11 @@
             this.setFacing = p_initialize_2_;
         }
     }
+    
+    //**Vivecraft Additions
+    
+	public void renderCustomModel(ModelResourceLocation model){
+		mc.getTextureManager().bindTexture(AtlasTexture.LOCATION_BLOCKS_TEXTURE);
+		mc.getBlockRendererDispatcher().getBlockModelRenderer().renderModelBrightnessColor(mc.modelManager.getModel(model), 1.0F, 1.0F, 1.0F, 1.0F);
+	}
 }
