--- a/net/minecraft/client/gui/screen/ConfirmBackupScreen.java
+++ b/net/minecraft/client/gui/screen/ConfirmBackupScreen.java
@@ -9,7 +9,7 @@
 
 public class ConfirmBackupScreen extends Screen
 {
-    private final Screen parentScreen;
+    private final Screen parentGuiScreen;
     protected final ConfirmBackupScreen.ICallback callback;
     private final ITextComponent message;
     private final boolean field_212994_d;
@@ -23,7 +23,7 @@
     public ConfirmBackupScreen(Screen p_i3824_1_, ConfirmBackupScreen.ICallback p_i3824_2_, ITextComponent p_i3824_3_, ITextComponent p_i3824_4_, boolean p_i3824_5_)
     {
         super(p_i3824_3_);
-        this.parentScreen = p_i3824_1_;
+        this.parentGuiScreen = p_i3824_1_;
         this.callback = p_i3824_2_;
         this.message = p_i3824_4_;
         this.field_212994_d = p_i3824_5_;
@@ -49,7 +49,7 @@
         }));
         this.addButton(new Button(this.width / 2 - 155 + 80, 124 + i, 150, 20, this.cancelText, (p_212991_1_) ->
         {
-            this.minecraft.displayGuiScreen(this.parentScreen);
+            this.minecraft.displayGuiScreen(this.parentGuiScreen);
         }));
         this.field_212996_j = new CheckboxButton(this.width / 2 - 155 + 80, 76 + i, 150, 20, this.field_212995_f, false);
 
@@ -83,7 +83,7 @@
     {
         if (p_keyPressed_1_ == 256)
         {
-            this.minecraft.displayGuiScreen(this.parentScreen);
+            this.minecraft.displayGuiScreen(this.parentGuiScreen);
             return true;
         }
         else
