--- a/net/minecraft/client/gui/screen/CreateWorldScreen.java
+++ b/net/minecraft/client/gui/screen/CreateWorldScreen.java
@@ -18,7 +18,7 @@
 
 public class CreateWorldScreen extends Screen
 {
-    private final Screen parentScreen;
+    private final Screen parentGuiScreen;
     private TextFieldWidget worldNameField;
     private TextFieldWidget worldSeedField;
     private String saveDirName;
@@ -49,7 +49,7 @@
     public CreateWorldScreen(Screen p_i1689_1_)
     {
         super(new TranslationTextComponent("selectWorld.create"));
-        this.parentScreen = p_i1689_1_;
+        this.parentGuiScreen = p_i1689_1_;
         this.worldSeed = "";
         this.worldName = I18n.format("selectWorld.newWorld");
     }
@@ -190,7 +190,7 @@
         }));
         this.addButton(new Button(this.width / 2 + 5, this.height - 28, 150, 20, I18n.format("gui.cancel"), (p_214317_1_) ->
         {
-            this.minecraft.displayGuiScreen(this.parentScreen);
+            this.minecraft.displayGuiScreen(this.parentGuiScreen);
         }));
         this.showMoreWorldOptions(this.inMoreWorldOptionsDisplay);
         this.func_212928_a(this.worldNameField);
