--- a/net/minecraft/client/gui/screen/EditSignScreen.java
+++ b/net/minecraft/client/gui/screen/EditSignScreen.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.gui.screen;
 
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+
 import com.mojang.blaze3d.platform.GlStateManager;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.StandingSignBlock;
@@ -35,6 +37,7 @@
             this.close();
         }));
         this.tileSign.setEditable(false);
+        KeyboardHandler.setOverlayShowing(true);
         this.field_214267_d = new TextInputUtil(this.minecraft, () ->
         {
             return this.tileSign.getText(this.editLine).getString();
@@ -55,6 +58,7 @@
         }
 
         this.tileSign.setEditable(true);
+        KeyboardHandler.setOverlayShowing(false);
     }
 
     public void tick()
