--- a/net/minecraft/client/gui/screen/FlatPresetsScreen.java
+++ b/net/minecraft/client/gui/screen/FlatPresetsScreen.java
@@ -30,7 +30,7 @@
 public class FlatPresetsScreen extends Screen
 {
     private static final List<FlatPresetsScreen.LayerItem> FLAT_WORLD_PRESETS = Lists.newArrayList();
-    private final CreateFlatWorldScreen parentScreen;
+    private final CreateFlatWorldScreen parentGuiScreen;
     private String presetsShare;
     private String listText;
     private FlatPresetsScreen.SlotList list;
@@ -40,7 +40,7 @@
     public FlatPresetsScreen(CreateFlatWorldScreen p_i1351_1_)
     {
         super(new TranslationTextComponent("createWorld.customize.presets.title"));
-        this.parentScreen = p_i1351_1_;
+        this.parentGuiScreen = p_i1351_1_;
     }
 
     protected void init()
@@ -50,18 +50,18 @@
         this.listText = I18n.format("createWorld.customize.presets.list");
         this.export = new TextFieldWidget(this.font, 50, 40, this.width - 100, 20, this.presetsShare);
         this.export.setMaxStringLength(1230);
-        this.export.setText(this.parentScreen.getPreset());
+        this.export.setText(this.parentGuiScreen.getPreset());
         this.children.add(this.export);
         this.list = new FlatPresetsScreen.SlotList();
         this.children.add(this.list);
         this.btnSelect = this.addButton(new Button(this.width / 2 - 155, this.height - 28, 150, 20, I18n.format("createWorld.customize.presets.select"), (p_213077_1_) ->
         {
-            this.parentScreen.setPreset(this.export.getText());
-            this.minecraft.displayGuiScreen(this.parentScreen);
+            this.parentGuiScreen.setPreset(this.export.getText());
+            this.minecraft.displayGuiScreen(this.parentGuiScreen);
         }));
         this.addButton(new Button(this.width / 2 + 5, this.height - 28, 150, 20, I18n.format("gui.cancel"), (p_213076_1_) ->
         {
-            this.minecraft.displayGuiScreen(this.parentScreen);
+            this.minecraft.displayGuiScreen(this.parentGuiScreen);
         }));
         this.func_213074_a(this.list.getSelected() != null);
     }
