--- a/net/minecraft/client/gui/screen/Screen.java
+++ b/net/minecraft/client/gui/screen/Screen.java
@@ -57,6 +57,11 @@
     protected FontRenderer font;
     private URI clickedLink;
 
+    public boolean pressShiftFake;
+    public static boolean isFakeShift(){
+    	return Minecraft.getInstance().currentScreen != null ? Minecraft.getInstance().currentScreen.pressShiftFake : false;
+    }
+    
     protected Screen(ITextComponent p_i402_1_)
     {
         this.title = p_i402_1_;
@@ -437,7 +442,16 @@
     {
         if (this.minecraft.world != null)
         {
-            this.fillGradient(0, 0, this.width, this.height, -1072689136, -804253680);
+        	//Vivecraft
+        	if (minecraft.vrSettings != null && minecraft.vrSettings.menuBackground == false)
+        	{
+        		this.fillGradient(0, 0, this.width, this.height, 0, 0);
+        	}
+        	else
+        	{
+        		this.fillGradient(0, 0, this.width, this.height, -1072689136, -804253680);
+        	}
+        	//
         }
         else
         {
@@ -497,7 +511,7 @@
 
     public static boolean hasShiftDown()
     {
-        return InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 344);
+        return InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 340) || InputMappings.isKeyDown(Minecraft.getInstance().mainWindow.getHandle(), 344) || isFakeShift();
     }
 
     public static boolean hasAltDown()
