--- a/net/minecraft/client/gui/screen/StatsScreen.java
+++ b/net/minecraft/client/gui/screen/StatsScreen.java
@@ -34,7 +34,7 @@
 
 public class StatsScreen extends Screen implements IProgressMeter
 {
-    protected final Screen parentScreen;
+    protected final Screen parentGuiScreen;
     private StatsScreen.CustomStatsList generalStats;
     private StatsScreen.StatsList itemStats;
     private StatsScreen.MobStatsList mobStats;
@@ -46,7 +46,7 @@
     public StatsScreen(Screen p_i2542_1_, StatisticsManager p_i2542_2_)
     {
         super(new TranslationTextComponent("gui.stats"));
-        this.parentScreen = p_i2542_1_;
+        this.parentGuiScreen = p_i2542_1_;
         this.stats = p_i2542_2_;
     }
 
@@ -79,7 +79,7 @@
         }));
         this.addButton(new Button(this.width / 2 - 100, this.height - 28, 200, 20, I18n.format("gui.done"), (p_213113_1_) ->
         {
-            this.minecraft.displayGuiScreen(this.parentScreen);
+            this.minecraft.displayGuiScreen(this.parentGuiScreen);
         }));
 
         if (this.itemStats.children().isEmpty())
