--- a/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
@@ -146,7 +146,7 @@
     {
         if (Minecraft.isGuiEnabled() && !entity.getDisplayedItem().isEmpty() && entity.getDisplayedItem().hasDisplayName() && this.renderManager.pointedEntity == entity)
         {
-            double d0 = entity.getDistanceSq(this.renderManager.info.getProjectedView());
+            double d0 = entity.getDistanceSq(this.renderManager.info.getPosition());
             float f = entity.func_213287_bg() ? 32.0F : 64.0F;
 
             if (d0 < (double)(f * f))
