--- a/net/minecraft/client/renderer/entity/MobRenderer.java
+++ b/net/minecraft/client/renderer/entity/MobRenderer.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.renderer.entity;
 
 import com.mojang.blaze3d.platform.GlStateManager;
+
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.BufferBuilder;
 import net.minecraft.client.renderer.Tessellator;
 import net.minecraft.client.renderer.culling.ICamera;
@@ -10,6 +12,7 @@
 import net.minecraft.entity.MobEntity;
 import net.minecraft.entity.item.HangingEntity;
 import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
 import net.optifine.Config;
 import net.optifine.shaders.Shaders;
 
@@ -77,6 +80,17 @@
                 double d7 = MathHelper.lerp((double)partialTicks, entity.prevPosY + (double)entity.getEyeHeight() * 0.7D, entity.posY + (double)entity.getEyeHeight() * 0.7D) - d4 * 0.5D - 0.25D;
                 double d8 = MathHelper.lerp((double)partialTicks, entity.prevPosZ, entity.posZ) - d3 * 0.7D + d2 * 0.5D * d5;
                 double d9 = (double)(MathHelper.lerp(partialTicks, entityLivingIn.renderYawOffset, entityLivingIn.prevRenderYawOffset) * ((float)Math.PI / 180F)) + (Math.PI / 2D);
+            
+                // VIVE START 
+                if (entity == Minecraft.getInstance().player)
+                {
+                	Vec3d aimSource = Minecraft.getInstance().vrPlayer.vrdata_world_render.getController(0).getPosition();
+                	d6 = aimSource.x;
+                	d7 = aimSource.y; 
+                	d8 = aimSource.z;             		         
+                }
+                // VIVE END 
+                
                 d2 = Math.cos(d9) * (double)entityLivingIn.getWidth() * 0.4D;
                 d3 = Math.sin(d9) * (double)entityLivingIn.getWidth() * 0.4D;
                 double d10 = MathHelper.lerp((double)partialTicks, entityLivingIn.prevPosX, entityLivingIn.posX) + d2;
