--- a/net/minecraft/client/util/InputMappings.java
+++ b/net/minecraft/client/util/InputMappings.java
@@ -3,6 +3,9 @@
 import com.google.common.collect.Maps;
 import java.util.Map;
 import java.util.Objects;
+
+import org.vivecraft.utils.InputSimulator;
+
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.resources.I18n;
 import net.minecraft.util.IntHashMap;
@@ -14,7 +17,7 @@
 
     public static boolean isKeyDown(int keyCode)
     {
-        return GLFW.glfwGetKey(Minecraft.getMinecraft().mainWindow.getHandle(), keyCode) == 1;
+        return GLFW.glfwGetKey(Minecraft.getMinecraft().mainWindow.getHandle(), keyCode) == 1 || InputSimulator.isKeyDown(keyCode);
     }
 
     public static InputMappings.Input getInputByCode(int keyCode, int scanCode)
