--- a/net/minecraft/entity/monster/EntityEnderman.java
+++ b/net/minecraft/entity/monster/EntityEnderman.java
@@ -6,6 +6,10 @@
 import java.util.function.Function;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.VivePlayer;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
@@ -192,7 +196,17 @@
         }
         else
         {
-            Vec3d vec3d = player.getLook(1.0F).normalize();
+        	//VIVE FIX
+            Vec3d vec3d;
+            VivePlayer vp = NetworkHelper.vivePlayers.get(player.getUniqueID());
+            if (vp !=null && vp.isVR())
+            {
+            	vec3d = vp.getHMDDir();
+            } else {
+            	vec3d = player.getLookVec();
+            }
+        	//
+        	
             Vec3d vec3d1 = new Vec3d(this.posX - player.posX, this.getEntityBoundingBox().minY + (double)this.getEyeHeight() - (player.posY + (double)player.getEyeHeight()), this.posZ - player.posZ);
             double d0 = vec3d1.length();
             vec3d1 = vec3d1.normalize();
