--- a/net/minecraft/util/datafix/fixes/LeavesFix.java
+++ b/net/minecraft/util/datafix/fixes/LeavesFix.java
@@ -362,7 +362,7 @@
             else
             {
                 Optional < List < Pair < String, Dynamic<? >>> > optional = p_i1792_1_.getOptional(this.paletteFinder);
-                this.palette = optional.map((p_208463_0_) ->
+                this.palette = (List)optional.map((p_208463_0_) ->
                 {
                     return p_208463_0_.stream().map(Pair::getSecond).collect(Collectors.toList());
                 }).orElse(ImmutableList.of());
