--- a/net/minecraft/client/gui/GuiScreen.java
+++ b/net/minecraft/client/gui/GuiScreen.java
@@ -62,7 +62,12 @@
     /** The FontRenderer used by GuiScreen */
     protected FontRenderer fontRenderer;
     private URI clickedLinkURI;
-
+    
+    public boolean pressShiftFake;
+    public static boolean isFakeShift(){
+    	return Minecraft.getMinecraft().currentScreen != null ? Minecraft.getMinecraft().currentScreen.pressShiftFake : false;
+    }
+    
     /**
      * Draws the screen and all the components in it.
      */
@@ -472,7 +477,16 @@
     {
         if (this.mc.world != null)
         {
-            this.drawGradientRect(0, 0, this.width, this.height, -1072689136, -804253680);
+        	/** MINECRIFT */
+        	if (this.mc.vrSettings != null && this.mc.vrSettings.menuBackground == false)
+        	{
+        		this.drawGradientRect(0, 0, this.width, this.height, 0, 0);
+        	}
+        	else
+        	{
+        		this.drawGradientRect(0, 0, this.width, this.height, -1072689136, -804253680);
+        	}
+        	/** END MINECRIFT */
         }
         else
         {
@@ -547,7 +561,7 @@
      */
     public static boolean isShiftKeyDown()
     {
-        return InputMappings.isKeyDown(340) || InputMappings.isKeyDown(344);
+        return InputMappings.isKeyDown(340) || InputMappings.isKeyDown(344) || isFakeShift();
     }
 
     /**
