--- a/net/minecraft/item/Item.java
+++ b/net/minecraft/item/Item.java
@@ -8,6 +8,10 @@
 import java.util.Random;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.VivePlayer;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.client.util.ITooltipFlag;
@@ -400,6 +404,22 @@
         double d0 = playerIn.posX;
         double d1 = playerIn.posY + (double)playerIn.getEyeHeight();
         double d2 = playerIn.posZ;
+        
+        //Vivecraft - Good Luck keeping this viable with Forge!
+    	VivePlayer vp = NetworkHelper.vivePlayers.get(playerIn.getUniqueID());
+    	if (vp !=null && vp.isVR())
+    	{
+    		Vec3d pos = vp.getControllerPos(0);
+    		d0 = pos.x;
+    		d1 = pos.y;
+    		d2 = pos.z;
+    		Vec3d aim = vp.getControllerDir(0);
+    		f = -(float)Math.toDegrees(Math.asin(aim.y/aim.length()));
+    		f1 = (float)Math.toDegrees(Math.atan2(-aim.x, aim.z));   		       		
+
+    	}           	         	
+        //
+        
         Vec3d vec3d = new Vec3d(d0, d1, d2);
         float f2 = MathHelper.cos(-f1 * 0.017453292F - (float)Math.PI);
         float f3 = MathHelper.sin(-f1 * 0.017453292F - (float)Math.PI);
