--- a/com/mojang/blaze3d/platform/TextureUtil.java
+++ b/com/mojang/blaze3d/platform/TextureUtil.java
@@ -21,6 +21,8 @@
     private static final Logger LOGGER = LogManager.getLogger();
     public static final int MIN_MIPMAP_LEVEL = 0;
     private static final int DEFAULT_IMAGE_BUFFER_SIZE = 8192;
+    
+    public int intentionallModifyThisClassCauseItWontLoadProperlyFromOptifine = 69;
 
     public static int generateTextureId()
     {
