--- a/net/minecraft/tileentity/TileEntityChest.java
+++ b/net/minecraft/tileentity/TileEntityChest.java
@@ -30,10 +30,10 @@
     private NonNullList<ItemStack> chestContents;
 
     /** The current angle of the lid (between 0 and 1) */
-    protected float lidAngle;
+    public float lidAngle;
 
     /** The angle of the lid last tick */
-    protected float prevLidAngle;
+    public float prevLidAngle;
 
     /** The number of players currently using this chest */
     protected int numPlayersUsing;
