--- a/net/minecraft/util/ScreenShotHelper.java
+++ b/net/minecraft/util/ScreenShotHelper.java
@@ -105,13 +105,13 @@
 
             file2 = (File)Reflector.call(object, Reflector.ScreenshotEvent_getScreenshotFile);
         }
-
+        final File file2f = file2;
         SimpleResource.RESOURCE_IO_EXECUTOR.execute(() ->
         {
             try {
-                nativeimage.write(file2);
-                ITextComponent itextcomponent1 = (new StringTextComponent(file2.getName())).applyTextStyle(TextFormatting.UNDERLINE).applyTextStyle((p_lambda$null$0_1_) -> {
-                    p_lambda$null$0_1_.setClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, file2.getAbsolutePath()));
+                nativeimage.write(file2f);
+                ITextComponent itextcomponent1 = (new StringTextComponent(file2f.getName())).applyTextStyle(TextFormatting.UNDERLINE).applyTextStyle((p_lambda$null$0_1_) -> {
+                    p_lambda$null$0_1_.setClickEvent(new ClickEvent(ClickEvent.Action.OPEN_FILE, file2f.getAbsolutePath()));
                 });
                 p_148259_5_.accept(new TranslationTextComponent("screenshot.success", itextcomponent1));
             }
