--- a/net/minecraft/client/gui/GuiChat.java
+++ b/net/minecraft/client/gui/GuiChat.java
@@ -16,6 +16,8 @@
 import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import com.mojang.brigadier.tree.CommandNode;
 import com.mojang.brigadier.tree.LiteralCommandNode;
+import com.mtbs3d.minecrift.gameplay.screenhandlers.KeyboardHandler;
+
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
@@ -88,6 +90,7 @@
         this.inputField = new GuiTextField(0, this.fontRenderer, 4, this.height - 12, this.width - 4, 12);
         this.inputField.setMaxStringLength(256);
         this.inputField.setEnableBackgroundDrawing(false);
+        this.inputField.setCanLoseFocus(false);
         this.inputField.setFocused(true);
         this.inputField.setText(this.defaultInputFieldText);
         this.inputField.setCanLoseFocus(false);
@@ -185,7 +188,12 @@
                 this.sendChatMessage(s);
             }
 
-            this.mc.displayGuiScreen((GuiScreen)null);
+            // VIVE - don't close chat if using SteamVR keyboard - is this still needed with new keyboard???
+            if (KeyboardHandler.Showing) {
+            	this.inputField.setText("");
+            } else {
+            	this.mc.displayGuiScreen((GuiScreen)null);
+            }
             return true;
         }
     }
